<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>danRerLib.KEGG &mdash; danrerlib  documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="danRerLib.enrichment" href="../enrichment/index.html" />
    <link rel="prev" title="danRerLib.GO" href="../GO/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            danrerlib
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../example.html">Example usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../conduct.html">Code of Conduct</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib</span></code></a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../index.html#subpackages">Subpackages</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../GO/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.GO</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.KEGG</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../enrichment/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.enrichment</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../mapping/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.mapping</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../settings/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.settings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../utils/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.utils</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#package-contents">Package Contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">danrerlib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.KEGG</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/autoapi/danRerLib/KEGG/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-danRerLib.KEGG">
<span id="danrerlib-kegg"></span><h1><a class="reference internal" href="#module-danRerLib.KEGG" title="danRerLib.KEGG"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.KEGG</span></code></a><a class="headerlink" href="#module-danRerLib.KEGG" title="Permalink to this heading"></a></h1>
<section id="kegg-module">
<p>This module provides functions for retrieving gene information associated with KEGG pathways and diseases
for various organisms, including human, zebrafish, and mapped zebrafish.</p>
<dl class="simple">
<dt>Functions:</dt><dd><ul class="simple">
<li><p><cite>get_genes_in_pathway</cite>: Retrieve genes associated with a specific KEGG pathway.</p></li>
<li><p><cite>get_genes_in_disease</cite>: Retrieve genes associated with a disease for a specified organism.</p></li>
</ul>
</dd>
<dt>Constants:</dt><dd><ul class="simple">
<li><p><cite>NCBI_ID</cite>: Identifier for ‘NCBI Gene ID’.</p></li>
<li><p><cite>ZFIN_ID</cite>: Identifier for ‘ZFIN ID’.</p></li>
<li><p><cite>ENS_ID</cite>: Identifier for ‘Ensembl ID’.</p></li>
<li><p><cite>SYMBOL</cite>: Identifier for gene ‘Symbol’.</p></li>
<li><p><cite>HUMAN_ID</cite>: Identifier for ‘Human NCBI Gene ID’.</p></li>
</ul>
</dd>
<dt>Database Rebuild Functions:</dt><dd><ul class="simple">
<li><p><cite>build_kegg_database</cite>: Build or update the KEGG pathway and disease database.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This module is designed for accessing gene information from KEGG pathways and diseases.</p></li>
<li><p>It provides functions to retrieve gene data associated with specific pathways and diseases.</p></li>
<li><p>Data is obtained either through direct API calls or by reading pre-processed files.</p></li>
<li><p>Organism options include ‘dre’ (Danio rerio), ‘hsa’ (Human), or ‘dreM’ (Mapped Danio rerio from Human).</p></li>
</ul>
<p class="rubric">Example</p>
<p>To retrieve genes associated with a specific KEGG pathway:
<code class="docutils literal notranslate"><span class="pre">`</span>
<span class="pre">pathway_genes</span> <span class="pre">=</span> <span class="pre">get_genes_in_pathway('hsa04010',</span> <span class="pre">'hsa')</span>
<span class="pre">`</span></code></p>
<p>For detailed information on each function and their usage, please refer to the documentation. For more examples of full functionality, please refer to tutorials.</p>
</section>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this heading"></a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG.get_genes_in_pathway" title="danRerLib.KEGG.get_genes_in_pathway"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_genes_in_pathway</span></code></a>(pathway_id[, org])</p></td>
<td><p>Retrieve genes associated with a specific pathway from KEGG.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG.get_genes_in_disease" title="danRerLib.KEGG.get_genes_in_disease"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_genes_in_disease</span></code></a>(→ pandas.DataFrame)</p></td>
<td><p>Retrieve genes associated with a disease for a specified organism.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG._check_for_organism" title="danRerLib.KEGG._check_for_organism"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_check_for_organism</span></code></a>(→ [str, str])</p></td>
<td><p>Check for and validate the organism information in a KEGG pathway ID.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG._check_if_pathway_id_exists" title="danRerLib.KEGG._check_if_pathway_id_exists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_check_if_pathway_id_exists</span></code></a>(→ bool)</p></td>
<td><p>Check if a KEGG pathway ID exists for the specified organism.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG.build_kegg_database" title="danRerLib.KEGG.build_kegg_database"><code class="xref py py-obj docutils literal notranslate"><span class="pre">build_kegg_database</span></code></a>()</p></td>
<td><p>Build or update the KEGG pathway and disease database.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG._download_zebrafish_pathway_ids" title="danRerLib.KEGG._download_zebrafish_pathway_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_download_zebrafish_pathway_ids</span></code></a>(zebrafish_pathways_path)</p></td>
<td><p>Download and store the list of zebrafish pathways and pathway names.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG._download_human_pathway_ids" title="danRerLib.KEGG._download_human_pathway_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_download_human_pathway_ids</span></code></a>(human_pathways_path)</p></td>
<td><p>Download and store the list of human pathways and pathway names.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG._create_mapped_zebrafish_pathway_ids" title="danRerLib.KEGG._create_mapped_zebrafish_pathway_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_create_mapped_zebrafish_pathway_ids</span></code></a>(...)</p></td>
<td><p>Create mapped zebrafish pathway IDs from human pathways and store them.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG._download_disease_ids" title="danRerLib.KEGG._download_disease_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_download_disease_ids</span></code></a>(human_disease_path)</p></td>
<td><p>Download and store the list of KEGG disease IDs and descriptions for humans.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG._download_human_pathway_genes" title="danRerLib.KEGG._download_human_pathway_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_download_human_pathway_genes</span></code></a>(human_pathways_path, ...)</p></td>
<td><p>Download and store genes associated with true KEGG pathways for humans.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG._download_zebrafish_pathway_genes" title="danRerLib.KEGG._download_zebrafish_pathway_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_download_zebrafish_pathway_genes</span></code></a>(...)</p></td>
<td><p>Download and store genes associated with true KEGG pathways for zebrafish.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG._download_human_disease_genes" title="danRerLib.KEGG._download_human_disease_genes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_download_human_disease_genes</span></code></a>(disease_ids_path)</p></td>
<td><p>Download and store genes associated with human diseases.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG._get_genes_for_disease" title="danRerLib.KEGG._get_genes_for_disease"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_get_genes_for_disease</span></code></a>(disease_id)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG._build_dre_mapped_pathway" title="danRerLib.KEGG._build_dre_mapped_pathway"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_build_dre_mapped_pathway</span></code></a>(human_pathways_path, ...)</p></td>
<td><p>Build mapped zebrafish KEGG pathways from human pathways and store them.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG._build_dre_mapped_disease" title="danRerLib.KEGG._build_dre_mapped_disease"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_build_dre_mapped_disease</span></code></a>(→ None)</p></td>
<td><p>Build mapped Zebrafish KEGG disease genes from human genes and store them.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG.testing" title="danRerLib.KEGG.testing"><code class="xref py py-obj docutils literal notranslate"><span class="pre">testing</span></code></a>()</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="attributes">
<h3>Attributes<a class="headerlink" href="#attributes" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG.zebrafish_pathways_path" title="danRerLib.KEGG.zebrafish_pathways_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">zebrafish_pathways_path</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG.human_pathways_path" title="danRerLib.KEGG.human_pathways_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">human_pathways_path</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG.mapped_zebrafish_pathways_path" title="danRerLib.KEGG.mapped_zebrafish_pathways_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mapped_zebrafish_pathways_path</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG.zebrafish_pathways_dir" title="danRerLib.KEGG.zebrafish_pathways_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">zebrafish_pathways_dir</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG.human_pathways_dir" title="danRerLib.KEGG.human_pathways_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">human_pathways_dir</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG.dre_mapped_dir" title="danRerLib.KEGG.dre_mapped_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dre_mapped_dir</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG.human_disease_path" title="danRerLib.KEGG.human_disease_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">human_disease_path</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.KEGG.human_disease_genes_path" title="danRerLib.KEGG.human_disease_genes_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">human_disease_genes_path</span></code></a></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.KEGG.dre_disease_genes_path" title="danRerLib.KEGG.dre_disease_genes_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dre_disease_genes_path</span></code></a></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.zebrafish_pathways_path">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">zebrafish_pathways_path</span></span><a class="headerlink" href="#danRerLib.KEGG.zebrafish_pathways_path" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.human_pathways_path">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">human_pathways_path</span></span><a class="headerlink" href="#danRerLib.KEGG.human_pathways_path" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.mapped_zebrafish_pathways_path">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">mapped_zebrafish_pathways_path</span></span><a class="headerlink" href="#danRerLib.KEGG.mapped_zebrafish_pathways_path" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.zebrafish_pathways_dir">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">zebrafish_pathways_dir</span></span><a class="headerlink" href="#danRerLib.KEGG.zebrafish_pathways_dir" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.human_pathways_dir">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">human_pathways_dir</span></span><a class="headerlink" href="#danRerLib.KEGG.human_pathways_dir" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.dre_mapped_dir">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">dre_mapped_dir</span></span><a class="headerlink" href="#danRerLib.KEGG.dre_mapped_dir" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.human_disease_path">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">human_disease_path</span></span><a class="headerlink" href="#danRerLib.KEGG.human_disease_path" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.human_disease_genes_path">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">human_disease_genes_path</span></span><a class="headerlink" href="#danRerLib.KEGG.human_disease_genes_path" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="danRerLib.KEGG.dre_disease_genes_path">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">dre_disease_genes_path</span></span><a class="headerlink" href="#danRerLib.KEGG.dre_disease_genes_path" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG.get_genes_in_pathway">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">get_genes_in_pathway</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pathway_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">org</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#get_genes_in_pathway"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG.get_genes_in_pathway" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve genes associated with a specific pathway from KEGG.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pathway_id</strong> (<em>str</em>) – The KEGG pathway ID for the pathway of interest.</p></li>
<li><p><strong>org</strong> (<em>str</em><em>, </em><em>optional</em>) – <p>The organism for which to retrieve pathway information.
Options:</p>
<blockquote>
<div><ul>
<li><p>’dre’: Danio rerio (Zebrafish) pathways from KEGG.</p></li>
<li><p>’hsa’: Human pathways from KEGG.</p></li>
<li><p>’dreM’: Mapped Danio rerio pathways from human.</p></li>
</ul>
<p>(Default is None, which will use the provided ‘org’ or ‘hsa’ if ‘org’ is None.)</p>
</div></blockquote>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame containing genes associated with the specified pathway.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function retrieves gene information associated with a specific KEGG pathway.</p></li>
<li><p>The ‘pathway_id’ parameter should be a valid KEGG pathway identifier.</p></li>
<li><p>The ‘org’ parameter specifies the organism for which to retrieve pathway information.</p></li>
<li><p>Organism options include ‘dre’ (Danio rerio), ‘hsa’ (Human), or ‘dreM’ (Mapped Danio rerio from Human).</p></li>
<li><p>Data is obtained from KEGG and may require internet access to download pathway information.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG.get_genes_in_disease">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">get_genes_in_disease</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">disease_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">org</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">API</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#get_genes_in_disease"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG.get_genes_in_disease" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve genes associated with a disease for a specified organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>disease_id</strong> (<em>str</em>) – The disease ID for which genes are to be retrieved.</p></li>
<li><p><strong>org</strong> (<em>str</em>) – The organism for which genes should be retrieved. Options:
- ‘hsa’: Returns human genes associated with the disease.
- ‘dre’ or ‘dreM’: Returns human genes mapped to Zebrafish genes (same result, as Zebrafish disease not characterized on KEGG).</p></li>
<li><p><strong>API</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use the KEGG REST API for retrieval. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame containing genes associated with the disease for the specified organism.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If ‘API’ is set to True, the function fetches gene information using the KEGG REST API.</p></li>
<li><p>If ‘API’ is False, the function reads gene information from pre-processed files.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._check_for_organism">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_check_for_organism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pathway_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">org</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_check_for_organism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._check_for_organism" title="Permalink to this definition"></a></dt>
<dd><p>Check for and validate the organism information in a KEGG pathway ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pathway_id</strong> (<em>str</em>) – The KEGG pathway ID, which may or may not include organism information.</p></li>
<li><p><strong>org</strong> (<em>str</em><em>, </em><em>optional</em>) – The specified organism code to validate against the pathway ID.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the validated organism code and the modified pathway ID.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the organism code in the pathway ID does not match the specified ‘org’ parameter.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This internal function is used to validate and extract organism information from a KEGG pathway ID.</p></li>
<li><p>If the pathway ID includes organism information, it validates it against the ‘org’ parameter.</p></li>
<li><p>If ‘org’ is None, it extracts and returns the organism code from the pathway ID.</p></li>
<li><p>Raises a ValueError if there is a mismatch between the organism code in the pathway ID and the ‘org’ parameter.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._check_if_pathway_id_exists">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_check_if_pathway_id_exists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pathway_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">org</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_check_if_pathway_id_exists"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._check_if_pathway_id_exists" title="Permalink to this definition"></a></dt>
<dd><p>Check if a KEGG pathway ID exists for the specified organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pathway_id</strong> (<em>str</em>) – The KEGG pathway ID to be validated.</p></li>
<li><p><strong>org</strong> (<em>str</em>) – The specified organism code (e.g., ‘dre’, ‘dreM’, or ‘hsa’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the pathway ID exists for the specified organism, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This internal function checks whether a given KEGG pathway ID exists for the specified organism.</p></li>
<li><p>It reads relevant pathway data files based on the organism code and checks if the pathway ID is present.</p></li>
<li><p>Returns True if the pathway ID exists, False otherwise.</p></li>
<li><p>Provides an error message if the pathway ID or organism code is invalid.</p></li>
<li><p>Organism code options include: ‘dre’ (Danio rerio), ‘dreM’ (Mapped Danio rerio from Human), or ‘hsa’ (Human).</p></li>
<li><p>Pathway IDs must be entered as strings if the organism prefix is omitted.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG.build_kegg_database">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">build_kegg_database</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#build_kegg_database"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG.build_kegg_database" title="Permalink to this definition"></a></dt>
<dd><p>Build or update the KEGG pathway and disease database.</p>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database creation or update and should be run only during version updates.</p></li>
<li><p>The process may take some time, so exercise caution when running it.</p></li>
<li><dl class="simple">
<dt>The function performs the following steps:</dt><dd><ol class="arabic simple">
<li><p>Downloads zebrafish pathway IDs and stores them in a specified file.</p></li>
<li><p>Downloads human pathway IDs and stores them in a specified file.</p></li>
<li><p>Creates mapped zebrafish pathway IDs from human pathways and stores them in a specified file.</p></li>
<li><p>Downloads genes associated with true KEGG pathways for zebrafish.</p></li>
<li><p>Downloads genes associated with true KEGG pathways for humans.</p></li>
<li><p>Builds mapped zebrafish KEGG pathways from human pathways and stores them in a specified directory.</p></li>
<li><p>Downloads KEGG disease IDs and stores them in a specified file.</p></li>
</ol>
</dd>
</dl>
</li>
<li><p>Running this function should be done carefully, as it involves downloading and processing data from KEGG.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._download_zebrafish_pathway_ids">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_download_zebrafish_pathway_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">zebrafish_pathways_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_download_zebrafish_pathway_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._download_zebrafish_pathway_ids" title="Permalink to this definition"></a></dt>
<dd><p>Download and store the list of zebrafish pathways and pathway names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>zebrafish_pathways_path</strong> (<em>str</em>) – The file path to save the zebrafish pathway data.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database building and should be run only during version updates.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._download_human_pathway_ids">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_download_human_pathway_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">human_pathways_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_download_human_pathway_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._download_human_pathway_ids" title="Permalink to this definition"></a></dt>
<dd><p>Download and store the list of human pathways and pathway names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>human_pathways_path</strong> (<em>str</em>) – The file path to save the human pathway data.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database building and should be run only during version updates.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._create_mapped_zebrafish_pathway_ids">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_create_mapped_zebrafish_pathway_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mapped_zebrafish_pathways_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">human_pathways_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_create_mapped_zebrafish_pathway_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._create_mapped_zebrafish_pathway_ids" title="Permalink to this definition"></a></dt>
<dd><p>Create mapped zebrafish pathway IDs from human pathways and store them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mapped_zebrafish_pathways_path</strong> (<em>str</em>) – The file path to save the mapped zebrafish pathway data.</p></li>
<li><p><strong>human_pathways_path</strong> (<em>str</em>) – The file path containing human pathway data.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database building and should be run only during version updates.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._download_disease_ids">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_download_disease_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">human_disease_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_download_disease_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._download_disease_ids" title="Permalink to this definition"></a></dt>
<dd><p>Download and store the list of KEGG disease IDs and descriptions for humans.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>human_disease_path</strong> (<em>str</em>) – The file path to save the human disease data.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database building and should be run only during version updates.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._download_human_pathway_genes">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_download_human_pathway_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">human_pathways_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">human_pathways_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_download_human_pathway_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._download_human_pathway_genes" title="Permalink to this definition"></a></dt>
<dd><p>Download and store genes associated with true KEGG pathways for humans.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>human_pathways_path</strong> (<em>str</em>) – The file path containing human pathway data.</p></li>
<li><p><strong>human_pathways_dir</strong> (<em>str</em>) – The directory to save gene data associated with human pathways.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database building and should be run only during version updates.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._download_zebrafish_pathway_genes">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_download_zebrafish_pathway_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">zebrafish_pathways_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zebrafish_pathways_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_download_zebrafish_pathway_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._download_zebrafish_pathway_genes" title="Permalink to this definition"></a></dt>
<dd><p>Download and store genes associated with true KEGG pathways for zebrafish.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>zebrafish_pathways_path</strong> (<em>str</em>) – The file path containing zebrafish pathway data.</p></li>
<li><p><strong>zebrafish_pathways_dir</strong> (<em>str</em>) – The directory to save gene data associated with zebrafish pathways.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database building and should be run only during version updates.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._download_human_disease_genes">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_download_human_disease_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">disease_ids_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_download_human_disease_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._download_human_disease_genes" title="Permalink to this definition"></a></dt>
<dd><p>Download and store genes associated with human diseases.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>disease_ids_path</strong> (<em>str</em>) – The file path to the list of disease IDs.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function reads a list of human disease IDs from the specified file.</p></li>
<li><p>It iterates through each disease ID and retrieves genes associated with the disease.</p></li>
<li><p>Genes are stored in a DataFrame with columns ‘Human NCBI Gene ID’ and ‘Disease ID’.</p></li>
<li><p>The resulting DataFrame is saved to a file named ‘human_disease_genes.txt’.</p></li>
<li><p>This function is intended for use in building or updating a database of human disease genes.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._get_genes_for_disease">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_get_genes_for_disease</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">disease_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_get_genes_for_disease"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._get_genes_for_disease" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._build_dre_mapped_pathway">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_build_dre_mapped_pathway</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">human_pathways_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">human_pathways_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dre_mapped_dir</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_build_dre_mapped_pathway"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._build_dre_mapped_pathway" title="Permalink to this definition"></a></dt>
<dd><p>Build mapped zebrafish KEGG pathways from human pathways and store them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>human_pathways_path</strong> (<em>str</em>) – The file path containing human pathway data.</p></li>
<li><p><strong>human_pathways_dir</strong> (<em>str</em>) – The directory containing gene data associated with human pathways.</p></li>
<li><p><strong>dre_mapped_dir</strong> (<em>str</em>) – The directory to save mapped zebrafish pathway data.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function is intended for database building and should be run only during version updates.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG._build_dre_mapped_disease">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">_build_dre_mapped_disease</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">human_disease_genes_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dre_disease_genes_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#_build_dre_mapped_disease"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG._build_dre_mapped_disease" title="Permalink to this definition"></a></dt>
<dd><p>Build mapped Zebrafish KEGG disease genes from human genes and store them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>human_disease_genes_path</strong> (<em>str</em>) – Path to the file containing human disease-associated genes.</p></li>
<li><p><strong>dre_disease_genes_path</strong> (<em>str</em>) – Path to the file where the mapped Zebrafish disease genes will be stored.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function takes a list of human genes associated with diseases, converts them to Zebrafish orthologs,
and stores the resulting genes in the specified file.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.KEGG.testing">
<span class="sig-prename descclassname"><span class="pre">danRerLib.KEGG.</span></span><span class="sig-name descname"><span class="pre">testing</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/KEGG.html#testing"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.KEGG.testing" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../GO/index.html" class="btn btn-neutral float-left" title="danRerLib.GO" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../enrichment/index.html" class="btn btn-neutral float-right" title="danRerLib.enrichment" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Ashley Schwartz.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>