{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform Enrichment\n",
    "\n",
    "Author: Ashley Schwartz\n",
    "\n",
    "Date: July 17, 2023\n",
    "\n",
    "## Purpose and Background\n",
    "\n",
    "This tutorial explains the process of conducting gene set enrichment testing, a method employed to assess predefined biologically relevant gene sets such as pathways or biological processes. The goal is to determine if these sets contain a higher number of significant genes from an experimental dataset compared to what would be expected by random chance.\n",
    "\n",
    "Given a dataset with differential gene expression for genes with associated p-values, we can determine the gene sets (concepts) that have significantly higher significance values than expected at random. This can be done by testing against all concept ids within a concept, or a few you might be interested in. \n",
    "\n",
    "A large limitation to a variety of designed enrichment methodologies is the minimal information for zebrafish. To overcome this, we have developed a new organism type, 'dreM' that is a mirror of the human 'hsa' organism with zebrafish genes. For example, the KEGG database has 355 annotated pathways for humans (hsa) at the time of writing this tutorial. Zebrafish (dre) have only 177. We have mapped all genes in the 355 hsa pathways to create 355 dreM pathways.\n",
    "\n",
    "If you would like to reference true zebrafish pathways, choose org = dre\n",
    "\n",
    "If you would like to reference mapped zebrafish pathways, choose org = dreM\n",
    "\n",
    "Some key definitions:\n",
    "\n",
    "| Term | Definition/Description |\n",
    "| - | - |\n",
    "| concept | a database resource such as KEGG or GO |\n",
    "| KEGG | defined as a concept https://www.kegg.jp/kegg/pathway.html | \n",
    "| KEGG pathway | pathway is a database for the KEGG concept |\n",
    "| KEGG disease | disease is a database for the KEGG concept | \n",
    "\n",
    "In this tutorial we will be using the following key elements:\n",
    "\n",
    "| Item | Desctiption  |\n",
    "|-|-|\n",
    "`data/test_data/TPP.txt` | A differential expression dataset containing Gene IDs, Log2FC values, and associated p-values |\n",
    "\n",
    "In general, while you do not need a large foundation in Python to execute the code listed in this tutorial, a general understanding of absolute and relative paths is useful.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up Python environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# IMPORT PYTHON PACKAGES\n",
    "# ----------------------\n",
    "\n",
    "# makes the notebook cell print all outputs\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = 'all'\n",
    "# path packages\n",
    "import sys\n",
    "from pathlib import Path\n",
    "# data processing packages\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SET UP MY LOCAL PACKAGE\n",
    "# -----------------------\n",
    "# this step is only needed because the local package has not been released through pip\n",
    "\n",
    "cwd = Path().absolute()\n",
    "\n",
    "package_folder = cwd / Path('../src/danRerlib')\n",
    "sys.path.append(str(package_folder))\n",
    "import mapping, KEGG, enrichment, utils\n",
    "\n",
    "# SET UP DATA DIRECTORY\n",
    "# ---------------------\n",
    "test_data_dir = cwd / Path('data/test_data/')\n",
    "out_data_dir = cwd / Path('data/out_data/')\n",
    "\n",
    "# note: I am using the Path package to take care of any operating\n",
    "#       system differences for users of this tutorial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find KEGG Enrichment \n",
    "\n",
    "### KEGG Pathway Enrichment\n",
    "\n",
    "_Purpose: Given a dataset with genes and associated p-values for significant differential expression, determine the enriched KEGG pathways._\n",
    "\n",
    "Step 1: Read your data into the workspace. The supported format is a dataset with columns 'NCBI Gene ID' and 'Pvalue'. We will be using the test dataset with relative path `data/test_data/TPP.txt` which is an example differential expression dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = test_data_dir / Path('TPP.txt')\n",
    "tpp_df = pd.read_csv(file_path, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also quickly print some stats to see what we are working with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "- The column names for this dataset are: ['NCBI Gene ID' 'PValue' 'logFC']\n",
      "- The data has 21854 entries (genes).\n"
     ]
    }
   ],
   "source": [
    "rows, cols = tpp_df.shape\n",
    "print(f\"- The column names for this dataset are: {tpp_df.columns.values}\")\n",
    "print(f'- The data has {rows} entries (genes).')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Step 2:__ Specify the Gene ID type currently used in your dataset. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id_type = 'NCBI Gene ID'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you'l notice, this matches my first column name in my dataset. A quick reminder that the Gene ID type must be one of the supported types and is case/spelling sensitive. Options are: NCBI Gene ID, ZFIN ID, Symbol, Ensembl ID. Many of the databases use the NCBI Gene ID so that is usually a preferred format, but any ID type will work here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Step 3:__ Launch the `enrich_KEGG` function. This function performs enrichment for the KEGG database with a few key distinctions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "out = enrichment.enrich_KEGG(tpp_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Concept Type</th>\n",
       "      <th>Concept ID</th>\n",
       "      <th># Genes in Concept in Universe</th>\n",
       "      <th># Sig Genes Belong to Concept</th>\n",
       "      <th>Proportion of Genes</th>\n",
       "      <th>Coeff</th>\n",
       "      <th>P-value</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Enriched</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM04550</td>\n",
       "      <td>155</td>\n",
       "      <td>47</td>\n",
       "      <td>0.303226</td>\n",
       "      <td>0.255060</td>\n",
       "      <td>4.961454e-07</td>\n",
       "      <td>4.961454e-07</td>\n",
       "      <td>1.290539</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM05225</td>\n",
       "      <td>173</td>\n",
       "      <td>56</td>\n",
       "      <td>0.323699</td>\n",
       "      <td>0.236088</td>\n",
       "      <td>2.403849e-06</td>\n",
       "      <td>2.403849e-06</td>\n",
       "      <td>1.266286</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM04961</td>\n",
       "      <td>64</td>\n",
       "      <td>18</td>\n",
       "      <td>0.281250</td>\n",
       "      <td>0.314233</td>\n",
       "      <td>2.836166e-06</td>\n",
       "      <td>2.836166e-06</td>\n",
       "      <td>1.369209</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM00030</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.379057</td>\n",
       "      <td>5.718199e-06</td>\n",
       "      <td>5.718199e-06</td>\n",
       "      <td>1.460907</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM04659</td>\n",
       "      <td>93</td>\n",
       "      <td>29</td>\n",
       "      <td>0.311828</td>\n",
       "      <td>0.275132</td>\n",
       "      <td>8.003806e-06</td>\n",
       "      <td>8.003806e-06</td>\n",
       "      <td>1.316705</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM04979</td>\n",
       "      <td>61</td>\n",
       "      <td>12</td>\n",
       "      <td>0.196721</td>\n",
       "      <td>0.183676</td>\n",
       "      <td>4.658255e-02</td>\n",
       "      <td>4.658255e-02</td>\n",
       "      <td>1.201626</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM04668</td>\n",
       "      <td>110</td>\n",
       "      <td>27</td>\n",
       "      <td>0.245455</td>\n",
       "      <td>0.147690</td>\n",
       "      <td>4.661458e-02</td>\n",
       "      <td>4.661458e-02</td>\n",
       "      <td>1.159154</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM05034</td>\n",
       "      <td>120</td>\n",
       "      <td>29</td>\n",
       "      <td>0.241667</td>\n",
       "      <td>0.142774</td>\n",
       "      <td>4.671743e-02</td>\n",
       "      <td>4.671743e-02</td>\n",
       "      <td>1.153469</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM04520</td>\n",
       "      <td>118</td>\n",
       "      <td>31</td>\n",
       "      <td>0.262712</td>\n",
       "      <td>0.142940</td>\n",
       "      <td>4.822297e-02</td>\n",
       "      <td>4.822297e-02</td>\n",
       "      <td>1.153660</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>KEGG pathway</td>\n",
       "      <td>dreM00830</td>\n",
       "      <td>29</td>\n",
       "      <td>7</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>0.233967</td>\n",
       "      <td>4.922382e-02</td>\n",
       "      <td>4.922382e-02</td>\n",
       "      <td>1.263603</td>\n",
       "      <td>enriched</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>122 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Concept Type Concept ID  # Genes in Concept in Universe  \\\n",
       "173  KEGG pathway  dreM04550                             155   \n",
       "272  KEGG pathway  dreM05225                             173   \n",
       "235  KEGG pathway  dreM04961                              64   \n",
       "10   KEGG pathway  dreM00030                              28   \n",
       "189  KEGG pathway  dreM04659                              93   \n",
       "..            ...        ...                             ...   \n",
       "230  KEGG pathway  dreM04979                              61   \n",
       "137  KEGG pathway  dreM04668                             110   \n",
       "335  KEGG pathway  dreM05034                             120   \n",
       "170  KEGG pathway  dreM04520                             118   \n",
       "83   KEGG pathway  dreM00830                              29   \n",
       "\n",
       "     # Sig Genes Belong to Concept  Proportion of Genes     Coeff  \\\n",
       "173                             47             0.303226  0.255060   \n",
       "272                             56             0.323699  0.236088   \n",
       "235                             18             0.281250  0.314233   \n",
       "10                              14             0.500000  0.379057   \n",
       "189                             29             0.311828  0.275132   \n",
       "..                             ...                  ...       ...   \n",
       "230                             12             0.196721  0.183676   \n",
       "137                             27             0.245455  0.147690   \n",
       "335                             29             0.241667  0.142774   \n",
       "170                             31             0.262712  0.142940   \n",
       "83                               7             0.241379  0.233967   \n",
       "\n",
       "          P-value           FDR  Odds Ratio  Enriched  \n",
       "173  4.961454e-07  4.961454e-07    1.290539  enriched  \n",
       "272  2.403849e-06  2.403849e-06    1.266286  enriched  \n",
       "235  2.836166e-06  2.836166e-06    1.369209  enriched  \n",
       "10   5.718199e-06  5.718199e-06    1.460907  enriched  \n",
       "189  8.003806e-06  8.003806e-06    1.316705  enriched  \n",
       "..            ...           ...         ...       ...  \n",
       "230  4.658255e-02  4.658255e-02    1.201626  enriched  \n",
       "137  4.661458e-02  4.661458e-02    1.159154  enriched  \n",
       "335  4.671743e-02  4.671743e-02    1.153469  enriched  \n",
       "170  4.822297e-02  4.822297e-02    1.153660  enriched  \n",
       "83   4.922382e-02  4.922382e-02    1.263603  enriched  \n",
       "\n",
       "[122 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinfo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
