<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>danrerlib.enrichment &mdash; danrerlib  documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/fonts.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/custom.css" type="text/css" />
    <link rel="canonical" href="https://sdsucomptox.github.io/danrerlib/autoapi/danrerlib/enrichment/index.html" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="danrerlib.enrichplots" href="../enrichplots/index.html" />
    <link rel="prev" title="danrerlib.KEGG" href="../KEGG/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            danrerlib
              <img src="../../../_static/danrerlib_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">User's Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../userguide/introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../userguide/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../userguide/usage.html">Usage Principles</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../GO/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib.GO</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../KEGG/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib.KEGG</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib.enrichment</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../enrichplots/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib.enrichplots</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../mapping/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib.mapping</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/perform_geneid_mapping.html">Gene ID Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/perform_orthology_mapping.html">Orthology Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/functional_enrichment.html">Functional Enrichment Analysis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Developer's Guide</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../devguide/changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../devguide/contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../devguide/conduct.html">Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../devguide/license.html">License</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">danrerlib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib.enrichment</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/autoapi/danrerlib/enrichment/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-danrerlib.enrichment">
<span id="danrerlib-enrichment"></span><h1><a class="reference internal" href="#module-danrerlib.enrichment" title="danrerlib.enrichment"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danrerlib.enrichment</span></code></a><a class="headerlink" href="#module-danrerlib.enrichment" title="Permalink to this heading"></a></h1>
<section id="enrichment-module">
<p>The Enrichment module offers a collection of functions to perform gene enrichment analyses. These analyses allow you to identify overrepresented gene sets or concepts in a given set of genes compared to a background or universe of genes. The module supports various enrichment databases, methods, and organisms, making it a versatile tool for uncovering biological insights.</p>
<dl>
<dt>Functions:</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">enrich</span></code>: Perform gene enrichment analysis using logistic regression or Fisher’s exact test for any of the supported annotation databases.</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>KEGG Pathway</cite></p></li>
<li><p><cite>KEGG Disease</cite></p></li>
<li><p><cite>GO BP</cite>: Gene Ontology Biological Processes</p></li>
<li><p><cite>GO CC</cite>: Gene Ontology Cellular Components</p></li>
<li><p><cite>GO MF</cite>: Gene Ontology Molecular Function</p></li>
<li><p><cite>GO</cite>: Gene Ontology Biological Processes, Cellular Components, and Molecular Function</p></li>
</ul>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">logistic</span></code>: Perform gene enrichment using logistic regression.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fishers</span></code>: Perform gene enrichment using Fisher’s exact test.</p></li>
</ul>
</dd>
<dt>Constants:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">NCBI_ID</span></code>: Identifier for NCBI Gene ID.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ZFIN_ID</span></code>: Identifier for ZFIN ID.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ENS_ID</span></code>: Identifier for Ensembl ID.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SYMBOL</span></code>: Identifier for gene Symbol.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">HUMAN_ID</span></code>: Identifier for Human NCBI Gene ID.</p></li>
</ul>
</dd>
<dt>Notes:</dt><dd><ul class="simple">
<li><p>The Enrichment module is designed for conducting gene enrichment analyses using various databases and methods.</p></li>
<li><p>It provides functions to analyze gene sets in the context of Gene Ontology (GO) and Kyoto Encyclopedia of Genes and Genomes (KEGG) concepts.</p></li>
<li><p>Users can choose from different gene ID types and organisms for analysis, enhancing flexibility.</p></li>
<li><p>The module includes statistical methods such as logistic regression and Fisher’s exact test for enrichment analysis.</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><p>To perform gene enrichment analysis for a set of zebrafish genes:</p>
<p><code class="docutils literal notranslate"><span class="pre">results</span> <span class="pre">=</span> <span class="pre">enrich(gene_universe,</span> <span class="pre">gene_id_type=ZFIN_ID,</span> <span class="pre">database='KEGG</span> <span class="pre">Pathway',</span> <span class="pre">method</span> <span class="pre">=</span> <span class="pre">'fishers',</span> <span class="pre">org='dre')</span></code></p>
<p>This example performs gene enrichment analysis using Fisher’s exact test on the specified gene set for zebrafish genes (org=’dre’).</p>
</dd>
</dl>
<p>For comprehensive details on each function and usage examples, please consult the documentation. You can also find tutorials demonstrating the full functionality of the Enrichment module.</p>
</section>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this heading"></a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#danrerlib.enrichment.enrich" title="danrerlib.enrichment.enrich"><code class="xref py py-obj docutils literal notranslate"><span class="pre">enrich</span></code></a>(→ danrerlib.settings.pd.DataFrame)</p></td>
<td><p>Perform gene enrichment analysis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danrerlib.enrichment.logistic" title="danrerlib.enrichment.logistic"><code class="xref py py-obj docutils literal notranslate"><span class="pre">logistic</span></code></a>(gene_universe_in, sig_genes_set, gene_set, ...)</p></td>
<td><p>Perform functional enrichment analysis using logistic regression.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danrerlib.enrichment.fishers" title="danrerlib.enrichment.fishers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fishers</span></code></a>(gene_universe, sig_genes_set, gene_set, ...)</p></td>
<td><p>Perform functional enrichment analysis using Fisher's exact test.</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="danrerlib.enrichment.enrich">
<span class="sig-prename descclassname"><span class="pre">danrerlib.enrichment.</span></span><span class="sig-name descname"><span class="pre">enrich</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_universe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">org</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dre'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logistic'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'both'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_gene_cutoff_pvalue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log2FC_cutoff_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concept_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_gene_set</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_conceptID_cutoff_pvalue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_by_p_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_num_genes_in_concept</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></span><a class="headerlink" href="#danrerlib.enrichment.enrich" title="Permalink to this definition"></a></dt>
<dd><p>Perform gene enrichment analysis.</p>
<dl>
<dt>Parameters:</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_universe</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame containing gene information, including gene IDs, p-values, and log2FC.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id_type</span> <span class="pre">(str)</span></code>: The type of gene ID in the gene universe. The recommended gene id type is NCBI Gene ID (NCBI_ID). Must be one of: NCBI Gene ID, ZFIN ID, Ensembl ID, Symbol, or for human: Human NCBI Gene ID.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">database</span> <span class="pre">(str</span> <span class="pre">or</span> <span class="pre">list[str])</span></code>: A list of functional annotation databases to test. Options include:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>KEGG Pathway</cite></p></li>
<li><p><cite>KEGG Disease</cite></p></li>
<li><p><cite>GO BP</cite>: Gene Ontology Biological Processes</p></li>
<li><p><cite>GO CC</cite>: Gene Ontology Cellular Components</p></li>
<li><p><cite>GO MF</cite>: Gene Ontology Molecular Function</p></li>
<li><p><cite>GO</cite>: Gene Ontology Biological Processes, Cellular Components, and Molecular Function</p></li>
<li><p><cite>all</cite>: all databases shown above</p></li>
<li><p>a list of any combination of the databases shown above. eg. databases = [KEGG Pathway, KEGG Disease]</p></li>
</ul>
</div></blockquote>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">org</span> <span class="pre">(str)</span></code>: The organism code (‘dre’ for zebrafish, ‘dreM’ for mapped zebrafish, ‘hsa’ for human).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">method</span> <span class="pre">(str,</span> <span class="pre">optional)</span></code>: The enrichment analysis method (‘logistic’ or ‘fishers’). Default is ‘logistic’.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">direction</span> <span class="pre">(str,</span> <span class="pre">optional)</span></code>: The direction of statistical test for enrichment (‘up’, ‘down’, or ‘both’). Default is ‘both’.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sig_gene_cutoff_pvalue</span> <span class="pre">(float,</span> <span class="pre">optional)</span></code>: The significance cutoff for gene inclusion based on p-values. Default is 0.05.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">log2FC_cutoff_value</span> <span class="pre">(float,</span> <span class="pre">optional)</span></code>: The log2 fold change cutoff value for gene inclusion. Default is 0.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">concept_ids</span> <span class="pre">(list,</span> <span class="pre">optional)</span></code>: A list of concept IDs (e.g., pathway IDs or disease IDs) to analyze. Default is None.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">background_gene_set</span> <span class="pre">(pd.DataFrame,</span> <span class="pre">optional)</span></code>: A DataFrame representing a background gene set. Default is None.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sig_conceptID_cutoff_pvalue</span> <span class="pre">(float,</span> <span class="pre">optional)</span></code>: The significance cutoff for concept IDs based on p-values. Default is 0.05.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">order_by_p_value</span> <span class="pre">(bool,</span> <span class="pre">optional)</span></code>: Whether to order the results by p-value. Default is True.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">min_num_genes_in_concept</span> <span class="pre">(int,</span> <span class="pre">optional)</span></code>: The minimum number of genes in a concept for it to be considered. Default is 10.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">include_all</span> <span class="pre">(bool,</span> <span class="pre">optional)</span></code>: Include all results without filtering based on significance. Default is False.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">result</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame containing enrichment analysis results, including concept details, the number of genes in the concept in the universe, the number of significant genes belonging to the concept, the proportion of genes in the concept, p-value, odds ratio, and enrichment direction.</p></li>
</ul>
</dd>
<dt>Note:</dt><dd><ul class="simple">
<li><p>If you are providing a list of concept ids, they must come from one database only.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danrerlib.enrichment.logistic">
<span class="sig-prename descclassname"><span class="pre">danrerlib.enrichment.</span></span><span class="sig-name descname"><span class="pre">logistic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_universe_in</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_genes_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concept_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concept_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concept_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_number_of_genes_in_universe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_direction</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#danrerlib.enrichment.logistic" title="Permalink to this definition"></a></dt>
<dd><p>Perform functional enrichment analysis using logistic regression.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gene_universe_in</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame representing the universe of genes.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sig_genes_set</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame containing the significantly expressed genes.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_set</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame containing the genes of interest.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id_type</span> <span class="pre">(str)</span></code>: The type of gene identifier used in the DataFrames.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">concept_type</span> <span class="pre">(str)</span></code>: The type of concept (e.g., pathway) being analyzed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">concept_id</span> <span class="pre">(str)</span></code>: The ID of the concept being analyzed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">concept_name</span> <span class="pre">(str)</span></code>: The name or description of the concept being analyzed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">total_number_of_genes_in_universe</span> <span class="pre">(int)</span></code>: The total number of genes in the universe.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">test_direction</span> <span class="pre">(str)</span></code>: The directionality of the test (‘greater’, ‘less’, or ‘two-sided’).</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">data</span> <span class="pre">(dict)</span></code>: A dictionary containing enrichment analysis results, including concept details,
the number of genes in the concept in the universe, the number of significant genes belonging
to the concept, the proportion of genes in the concept, odds ratio, p-value, and enrichment direction.</p></li>
</ul>
</dd>
<dt>Notes:</dt><dd><ul class="simple">
<li><p>This function performs functional enrichment analysis using logistic regression.</p></li>
<li><p>It calculates enrichment statistics for a specified concept (e.g., pathway) by comparing a gene set
of interest to a larger gene universe.</p></li>
<li><p>The ‘gene_id_type’ parameter specifies the type of gene identifiers used in the DataFrames.</p></li>
<li><p>The ‘test_direction’ parameter determines the directionality of the test (‘greater’, ‘less’, or ‘two-sided’).</p></li>
<li><p>Enrichment results include the odds ratio, p-value, and enrichment direction.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danrerlib.enrichment.fishers">
<span class="sig-prename descclassname"><span class="pre">danrerlib.enrichment.</span></span><span class="sig-name descname"><span class="pre">fishers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_universe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_genes_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">danrerlib.settings.pd.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concept_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concept_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concept_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_number_of_genes_in_universe</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_direction</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#danrerlib.enrichment.fishers" title="Permalink to this definition"></a></dt>
<dd><p>Perform functional enrichment analysis using Fisher’s exact test.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gene_universe</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame representing the universe of genes.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sig_genes_set</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame containing the significantly expressed genes.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_set</span> <span class="pre">(pd.DataFrame)</span></code>: A DataFrame containing the genes of interest.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gene_id_type</span> <span class="pre">(str)</span></code>: The type of gene identifier used in the DataFrames.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">concept_type</span> <span class="pre">(str)</span></code>: The type of concept (e.g., pathway) being analyzed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">concept_id</span> <span class="pre">(str)</span></code>: The ID of the concept being analyzed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">concept_name</span> <span class="pre">(str)</span></code>: The name or description of the concept being analyzed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">total_number_of_genes_in_universe</span> <span class="pre">(int)</span></code>: The total number of genes in the universe.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">test_direction</span> <span class="pre">(str)</span></code>: The directionality of the test.
Options: ‘two-sided’, ‘greater’, ‘less’.</p></li>
</ul>
</dd>
<dt>Returns:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">data</span> <span class="pre">(dict)</span></code>: A dictionary containing enrichment analysis results, including concept details,
the number of genes in the concept in the universe, the number of significant genes belonging
to the concept, the proportion of genes in the concept, odds ratio, p-value, and enrichment direction.</p></li>
</ul>
</dd>
<dt>Notes:</dt><dd><ul class="simple">
<li><p>This function performs functional enrichment analysis using Fisher’s exact test.</p></li>
<li><p>It calculates enrichment statistics for a specified concept (e.g., pathway) by comparing a gene set
of interest to a larger gene universe.</p></li>
<li><p>The ‘gene_id_type’ parameter specifies the type of gene identifiers used in the DataFrames.</p></li>
<li><p>The ‘test_direction’ parameter determines the directionality of the test (‘two-sided’, ‘greater’, ‘less’).</p></li>
<li><p>Enrichment results include the odds ratio, p-value, and enrichment direction.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../KEGG/index.html" class="btn btn-neutral float-left" title="danrerlib.KEGG" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../enrichplots/index.html" class="btn btn-neutral float-right" title="danrerlib.enrichplots" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Ashley Schwartz.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>