<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>danRerLib.mapping &mdash; danrerlib  documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="danRerLib.settings" href="../settings/index.html" />
    <link rel="prev" title="danRerLib.enrichment" href="../enrichment/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            danrerlib
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../example.html">Example usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../conduct.html">Code of Conduct</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib</span></code></a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../index.html#subpackages">Subpackages</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../GO/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.GO</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../KEGG/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.KEGG</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../enrichment/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.enrichment</span></code></a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.mapping</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../settings/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.settings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../utils/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.utils</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../index.html#package-contents">Package Contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">danrerlib</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.mapping</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/autoapi/danRerLib/mapping/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-danRerLib.mapping">
<span id="danrerlib-mapping"></span><h1><a class="reference internal" href="#module-danRerLib.mapping" title="danRerLib.mapping"><code class="xref py py-mod docutils literal notranslate"><span class="pre">danRerLib.mapping</span></code></a><a class="headerlink" href="#module-danRerLib.mapping" title="Permalink to this heading"></a></h1>
<section id="gene-mapping-module">
<p>This module provides functions for gene ID mapping and orthology checks between different species,
with a focus on Zebrafish and human gene IDs.</p>
<dl class="simple">
<dt>Functions:</dt><dd><ul class="simple">
<li><p><cite>convert_ids</cite>: Convert a list of Gene IDs between different types.</p></li>
<li><p><cite>add_mapped_column</cite>: Add a new column with mapped Gene IDs to a pandas DataFrame.</p></li>
<li><p><cite>convert_to_human</cite>: Convert Zebrafish gene IDs to their corresponding human orthologs.</p></li>
<li><p><cite>convert_to_zebrafish</cite>: Convert human gene IDs to their corresponding Zebrafish orthologs.</p></li>
<li><p><cite>add_mapped_ortholog_column</cite>: Add a new column with mapped ortholog Gene IDs to a pandas DataFrame.</p></li>
</ul>
</dd>
<dt>Constants:</dt><dd><ul class="simple">
<li><p><cite>NCBI_ID</cite>: Identifier for ‘NCBI Gene ID’.</p></li>
<li><p><cite>ZFIN_ID</cite>: Identifier for ‘ZFIN ID’.</p></li>
<li><p><cite>ENS_ID</cite>: Identifier for ‘Ensembl ID’.</p></li>
<li><p><cite>SYMBOL</cite>: Identifier for gene ‘Symbol’.</p></li>
<li><p><cite>HUMAN_ID</cite>: Identifier for ‘Human NCBI Gene ID’.</p></li>
</ul>
</dd>
<dt>Database Rebuild Functions:</dt><dd><ul class="simple">
<li><p><cite>build_gene_mapping</cite>: Build a master gene mapping file based on Zebrafish gene ID conversions.</p></li>
<li><p><cite>build_ortho_mapping</cite>: Build a master orthology mapping file between Zebrafish and human genes.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This module is designed for use in genomics research and gene ID mapping projects.</p></li>
<li><p>It provides functions to convert gene IDs between different types, build mapping files, and perform orthology checks.</p></li>
<li><p>If you would like to re-build mapping files, ensure that your dataset and mapping files are accurately named and organized as per the function requirements. It is recommended to use the current package database build.</p></li>
</ul>
<p class="rubric">Example</p>
<p>To convert a list of Zebrafish gene IDs to human orthologs:
<code class="docutils literal notranslate"><span class="pre">`</span>
<span class="pre">gene_list</span> <span class="pre">=</span> <span class="pre">['ENSDARG00000012345',</span> <span class="pre">'ENSDARG00000067890']</span>
<span class="pre">converted_genes</span> <span class="pre">=</span> <span class="pre">convert_to_human(gene_list,</span> <span class="pre">ENS_ID,</span> <span class="pre">HUMAN_ID)</span>
<span class="pre">`</span></code></p>
<p>For detailed information on each function and their usage, please refer to the documentation. For more examples of full functionality, please refer to tutorials.</p>
</section>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this heading"></a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.mapping.convert_ids" title="danRerLib.mapping.convert_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">convert_ids</span></code></a>(gene_list, id_from, id_to[, keep_mapping, ...])</p></td>
<td><p>Convert a list of Gene IDs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.mapping.add_mapped_column" title="danRerLib.mapping.add_mapped_column"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_mapped_column</span></code></a>(data, id_from, id_to[, ...])</p></td>
<td><p>Add a new column to a pandas DataFrame with mapped Gene IDs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.mapping.convert_to_human" title="danRerLib.mapping.convert_to_human"><code class="xref py py-obj docutils literal notranslate"><span class="pre">convert_to_human</span></code></a>(gene_list, zfish_gene_type[, ...])</p></td>
<td><p>Convert a list of Zebrafish gene IDs to their human orthologs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.mapping.convert_to_zebrafish" title="danRerLib.mapping.convert_to_zebrafish"><code class="xref py py-obj docutils literal notranslate"><span class="pre">convert_to_zebrafish</span></code></a>(gene_list, zfish_gene_type[, ...])</p></td>
<td><p>Convert a list of human gene IDs to their Zebrafish orthologs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.mapping.get_ortho_ids" title="danRerLib.mapping.get_ortho_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_ortho_ids</span></code></a>(gene_list, id_from, id_to[, ...])</p></td>
<td><p>Retrieve orthologous gene IDs for a given list of genes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.mapping.add_mapped_ortholog_column" title="danRerLib.mapping.add_mapped_ortholog_column"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_mapped_ortholog_column</span></code></a>(data, id_from, id_to[, ...])</p></td>
<td><p>Add a new column to a pandas DataFrame with mapped ortholog Gene IDs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.mapping._build_gene_mapping" title="danRerLib.mapping._build_gene_mapping"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_build_gene_mapping</span></code></a>()</p></td>
<td><p>Build the master gene mapping file named 'master_gene_mapping_file_V&lt;VERSION_NUM&gt;.txt'.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.mapping._build_ortho_mapping" title="danRerLib.mapping._build_ortho_mapping"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_build_ortho_mapping</span></code></a>()</p></td>
<td><p>Build the master orthology mapping file named 'master_ortho_mapping_file_V&lt;VERSION_NUM&gt;.txt'.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.mapping._make_sure_is_pandas_series" title="danRerLib.mapping._make_sure_is_pandas_series"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_make_sure_is_pandas_series</span></code></a>(gene_list, id_from)</p></td>
<td><p>Ensure that the input is a pandas Series.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.mapping._check_valid_zebrafish_gene_id_type" title="danRerLib.mapping._check_valid_zebrafish_gene_id_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_check_valid_zebrafish_gene_id_type</span></code></a>(gene_id_types)</p></td>
<td><p>Check the validity of Zebrafish gene ID types.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#danRerLib.mapping._check_valid_gene_id_type_for_orthology" title="danRerLib.mapping._check_valid_gene_id_type_for_orthology"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_check_valid_gene_id_type_for_orthology</span></code></a>(gene_id_from, ...)</p></td>
<td><p>Check the validity of gene ID types for orthology mapping.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#danRerLib.mapping._check_column_name_matches_id_choice" title="danRerLib.mapping._check_column_name_matches_id_choice"><code class="xref py py-obj docutils literal notranslate"><span class="pre">_check_column_name_matches_id_choice</span></code></a>(data, id_from, ...)</p></td>
<td><p>Check if the specified column name matches the chosen gene ID type.</p></td>
</tr>
</tbody>
</table>
<dl class="py exception">
<dt class="sig sig-object py" id="danRerLib.mapping.DatabaseNotFoundError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">DatabaseNotFoundError</span></span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#DatabaseNotFoundError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.DatabaseNotFoundError" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-obj docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Raised when the the database directory is not found</p>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="danRerLib.mapping.InvalidGeneTypeError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">InvalidGeneTypeError</span></span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#InvalidGeneTypeError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.InvalidGeneTypeError" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-obj docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Raised when the Gene ID Type is invalid</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping.convert_ids">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">convert_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_to</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#convert_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.convert_ids" title="Permalink to this definition"></a></dt>
<dd><p>Convert a list of Gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_list</strong> (<em>array-like</em>) – A list of Gene IDs with supported formats like list, pd.Series,
pd.DataFrame, or np.array.</p></li>
<li><p><strong>id_from</strong> (<em>str</em>) – The current Gene ID type.</p></li>
<li><p><strong>id_to</strong> (<em>str</em>) – The Gene ID type to convert to.</p></li>
<li><p><strong>keep_mapping</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to keep a mapping of the original IDs to the
converted IDs. Default is False.</p></li>
<li><p><strong>out_format</strong> (<em>str</em><em>, </em><em>optional</em>) – The desired output format for the result. Default is None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas Series or DataFrame containing the converted Gene IDs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series or pd.DataFrame</p>
</dd>
</dl>
<dl class="simple">
<dt>Other:</dt><dd><p>Gene ID Type Options: NCBI Gene ID, ZFIN ID, Symbol, Ensembl ID</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>if the order is important, it is recommended to have keep_mapping = True. Otherwise,
the mapping order is not guaranteed.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping.add_mapped_column">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">add_mapped_column</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_to</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_name_with_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_old_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#add_mapped_column"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.add_mapped_column" title="Permalink to this definition"></a></dt>
<dd><p>Add a new column to a pandas DataFrame with mapped Gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em><em>, </em><em>list</em>) – A pandas DataFrame containing a column that has Gene IDs of some type.</p></li>
<li><p><strong>id_from</strong> (<em>str</em>) – The current Gene ID type. Must be one of: NCBI Gene ID, ZFIN ID, Ensembl ID, or Symbol.</p></li>
<li><p><strong>id_to</strong> (<em>str</em>) – The Gene ID type to convert to. Must be one of: NCBI Gene ID, ZFIN ID, Ensembl ID, or Symbol.</p></li>
<li><p><strong>column_name_with_ids</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the column containing the Gene IDs if it doesn’t match id_from.</p></li>
<li><p><strong>keep_old_ids</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to keep the old Gene ID column. Default is True.</p></li>
<li><p><strong>drop_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to drop rows with NA values in the resulting mapped column. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame containing the added mapped column.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function adds a new column to the input DataFrame containing the mapped Gene IDs.</p></li>
<li><p>The new column will have the name ‘Gene ID (Mapped)’ unless specified otherwise.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping.convert_to_human">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">convert_to_human</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zfish_gene_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_missing_orthos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#convert_to_human"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.convert_to_human" title="Permalink to this definition"></a></dt>
<dd><p>Convert a list of Zebrafish gene IDs to their human orthologs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_list</strong> (<em>list</em>) – A list of Zebrafish gene IDs to be converted to human orthologs.</p></li>
<li><p><strong>zfish_gene_type</strong> (<em>str</em>) – The current gene ID type for the Zebrafish genes. Must be one of: NCBI Gene ID, ZFIN ID, Ensembl ID, or Symbol.</p></li>
<li><p><strong>keep_mapping</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to retain the mapping information. Default is False.</p></li>
<li><p><strong>keep_missing_orthos</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to keep gene IDs with missing orthologs. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of human gene IDs corresponding to the Zebrafish gene IDs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function converts Zebrafish gene IDs to their human orthologs using orthology mapping.</p></li>
<li><p>The <cite>zfish_gene_type</cite> parameter specifies the type of Zebrafish gene IDs.</p></li>
<li><p>To retain the mapping information, set <cite>keep_mapping</cite> to True.</p></li>
<li><p>To keep Zebrafish gene IDs with missing orthologs, set <cite>keep_missing_orthos</cite> to True.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping.convert_to_zebrafish">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">convert_to_zebrafish</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zfish_gene_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_missing_orthos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#convert_to_zebrafish"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.convert_to_zebrafish" title="Permalink to this definition"></a></dt>
<dd><p>Convert a list of human gene IDs to their Zebrafish orthologs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_list</strong> (<em>list</em>) – A list of human gene IDs to be converted to Zebrafish orthologs.</p></li>
<li><p><strong>zfish_gene_type</strong> (<em>str</em>) – The target gene ID type for the Zebrafish orthologs. Must be one of: NCBI Gene ID, ZFIN ID, Ensembl ID, or Symbol.</p></li>
<li><p><strong>keep_mapping</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to retain the mapping information. Default is False.</p></li>
<li><p><strong>keep_missing_orthos</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to keep gene IDs with missing orthologs. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of Zebrafish gene IDs corresponding to the human gene IDs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function converts human gene IDs to their Zebrafish orthologs using orthology mapping.</p></li>
<li><p>The <cite>zfish_gene_type</cite> parameter specifies the type of Zebrafish gene IDs.</p></li>
<li><p>To retain the mapping information, set <cite>keep_mapping</cite> to True.</p></li>
<li><p>To keep human gene IDs with missing Zebrafish orthologs, set <cite>keep_missing_orthos</cite> to True.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping.get_ortho_ids">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">get_ortho_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_to</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_mapping</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_missing_orthos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#get_ortho_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.get_ortho_ids" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve orthologous gene IDs for a given list of genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_list</strong> (<em>list</em>) – A list of gene IDs.</p></li>
<li><p><strong>id_from</strong> (<em>str</em>) – The current gene ID type. Must be one of: NCBI Gene ID, ZFIN ID, Ensembl ID, Symbol, or Human NCBI Gene ID.</p></li>
<li><p><strong>id_to</strong> (<em>str</em>) – The target gene ID type to convert to. Must be one of: NCBI Gene ID, ZFIN ID, Ensembl ID, Symbol, or Human NCBI Gene ID.</p></li>
<li><p><strong>keep_mapping</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to retain the mapping information. Default is False.</p></li>
<li><p><strong>keep_missing_orthos</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to keep gene IDs with missing orthologs. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of orthologous gene IDs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function retrieves orthologous gene IDs for the provided gene list.</p></li>
<li><p>The mapping information can be retained by setting <cite>keep_mapping</cite> to True.</p></li>
<li><p>Gene IDs with missing orthologs can be retained by setting <cite>keep_missing_orthos</cite> to True.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping.add_mapped_ortholog_column">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">add_mapped_ortholog_column</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_from</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_to</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_name_with_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_old_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_na</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#add_mapped_ortholog_column"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping.add_mapped_ortholog_column" title="Permalink to this definition"></a></dt>
<dd><p>Add a new column to a pandas DataFrame with mapped ortholog Gene IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – A pandas DataFrame containing a column with Gene IDs of the specified ‘id_from’ type.</p></li>
<li><p><strong>id_from</strong> (<em>str</em>) – The current Gene ID type to convert from. Must be one of: NCBI Gene ID, ZFIN ID,
Ensembl ID, Symbol, or Human NCBI Gene ID.</p></li>
<li><p><strong>id_to</strong> (<em>str</em>) – The Gene ID type to convert to. Must be one of: NCBI Gene ID, ZFIN ID,
Ensembl ID, Symbol, or Human NCBI Gene ID.</p></li>
<li><p><strong>column_name_with_ids</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the column containing the Gene IDs if it doesn’t match ‘id_from’.</p></li>
<li><p><strong>keep_old_ids</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to keep the original Gene ID column. Default is True.</p></li>
<li><p><strong>drop_na</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to drop rows with NA values in the resulting mapped column. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame containing the added mapped ortholog Gene IDs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function adds a new column to the input DataFrame containing the mapped ortholog Gene IDs.</p></li>
<li><p>The new column will have the name ‘Ortholog Gene ID (Mapped)’ unless specified otherwise.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping._build_gene_mapping">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">_build_gene_mapping</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#_build_gene_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping._build_gene_mapping" title="Permalink to this definition"></a></dt>
<dd><p>Build the master gene mapping file named ‘master_gene_mapping_file_V&lt;VERSION_NUM&gt;.txt’.</p>
<p>This function reads data files from the database/raw_data subdirectory to create the mapping file.
The required data files include:
- ‘zfin_to_ncbi_V&lt;VERSION_NUM&gt;.txt’
- ‘zfin_to_ensembl_V&lt;VERSION_NUM&gt;.txt’</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If one or both of the required data files are not found in the raw_data folder
of the database directory.</p></li>
<li><p><a class="reference internal" href="#danRerLib.mapping.DatabaseNotFoundError" title="danRerLib.mapping.DatabaseNotFoundError"><strong>DatabaseNotFoundError</strong></a> – If the database directory is not found in the current working directory.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function merges the ‘zfin_to_ncbi’ and ‘zfin_to_ensembl’ datasets, removing unnecessary columns.</p></li>
<li><p>The resulting merged DataFrame is saved as ‘master_gene_mapping_file_V&lt;VERSION_NUM&gt;.txt’ in the database directory.</p></li>
<li><p>Make sure to have the required data files in the raw_data subdirectory before running this function.</p></li>
</ul>
<p class="rubric">Example</p>
<p>To build the gene mapping file, you need the following data files:
- ‘zfin_to_ncbi_V&lt;VERSION_NUM&gt;.txt’
- ‘zfin_to_ensembl_V&lt;VERSION_NUM&gt;.txt’
Place them in the raw_data folder of the database directory.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping._build_ortho_mapping">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">_build_ortho_mapping</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#_build_ortho_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping._build_ortho_mapping" title="Permalink to this definition"></a></dt>
<dd><p>Build the master orthology mapping file named ‘master_ortho_mapping_file_V&lt;VERSION_NUM&gt;.txt’.</p>
<p>This function reads data files from the database/raw_data subdirectory to create the orthology mapping file.
The required data file includes:
- ‘zfish_human_orthology_V&lt;VERSION_NUM&gt;.txt’</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If the required data file is not found in the raw_data folder of the database directory.</p></li>
<li><p><a class="reference internal" href="#danRerLib.mapping.DatabaseNotFoundError" title="danRerLib.mapping.DatabaseNotFoundError"><strong>DatabaseNotFoundError</strong></a> – If the database directory is not found in the current working directory.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function processes the ‘zfish_human_orthology’ dataset, removing unnecessary columns.</p></li>
<li><p>Duplicate entries are removed, and the DataFrame is saved as ‘master_ortho_mapping_file_V&lt;VERSION_NUM&gt;.txt’
in the database directory.</p></li>
<li><p>Ensure that you have the required data file in the raw_data folder of the database directory before running this function.</p></li>
</ul>
<p class="rubric">Example</p>
<p>To build the orthology mapping file, you need the following data file:
- ‘zfish_human_orthology_V&lt;VERSION_NUM&gt;.txt’
Place it in the raw_data folder of the database directory.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping._make_sure_is_pandas_series">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">_make_sure_is_pandas_series</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_from</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#_make_sure_is_pandas_series"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping._make_sure_is_pandas_series" title="Permalink to this definition"></a></dt>
<dd><p>Ensure that the input is a pandas Series.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_list</strong> (<em>pd.Series</em><em>, </em><em>pd.DataFrame</em><em>, </em><em>list</em>) – Input data, which can be a Series, DataFrame, or list.</p></li>
<li><p><strong>id_from</strong> (<em>str</em>) – Identifier for the data.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas Series based on the input data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.Series</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function checks the type of the input data and converts it into a pandas Series if necessary.
If the input is a DataFrame, it will be squeezed into a Series.
If the input is a list, it will be converted into a Series with the specified name.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping._check_valid_zebrafish_gene_id_type">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">_check_valid_zebrafish_gene_id_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_id_types</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#_check_valid_zebrafish_gene_id_type"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping._check_valid_zebrafish_gene_id_type" title="Permalink to this definition"></a></dt>
<dd><p>Check the validity of Zebrafish gene ID types.</p>
<p>This function checks if the provided Zebrafish gene ID types are valid options.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>gene_id_types</strong> (<em>str</em><em> or </em><em>list</em>) – A string or a list of Zebrafish gene ID types to be validated.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#danRerLib.mapping.InvalidGeneTypeError" title="danRerLib.mapping.InvalidGeneTypeError"><strong>InvalidGeneTypeError</strong></a> – If one or more of the provided gene ID types are invalid.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>Valid Zebrafish gene ID types include: NCBI Gene ID, ZFIN ID, Ensembl ID, or Symbol.</p></li>
<li><p>The input can be a single gene ID type as a string or multiple types in a list.</p></li>
<li><p>Gene ID types are case and spelling sensitive.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping._check_valid_gene_id_type_for_orthology">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">_check_valid_gene_id_type_for_orthology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_id_from</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_to</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#_check_valid_gene_id_type_for_orthology"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping._check_valid_gene_id_type_for_orthology" title="Permalink to this definition"></a></dt>
<dd><p>Check the validity of gene ID types for orthology mapping.</p>
<p>This function checks if the provided gene ID types are valid options for orthology mapping.
It ensures that the gene ID types are not from the same organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_id_from</strong> (<em>str</em>) – The source gene ID type.</p></li>
<li><p><strong>gene_id_to</strong> (<em>str</em>) – The target gene ID type.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#danRerLib.mapping.InvalidGeneTypeError" title="danRerLib.mapping.InvalidGeneTypeError"><strong>InvalidGeneTypeError</strong></a> – If one or both of the provided gene ID types are invalid or from the same organism.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>Valid Zebrafish gene ID types include: NCBI_ID, ZFIN_ID, ENS_ID, and SYMBOL.</p></li>
<li><p>Valid human gene ID type is: HUMAN_ID.</p></li>
<li><p>Gene ID types are case and spelling sensitive.</p></li>
<li><p>To perform orthology mapping, id_from must be from organism A, and id_to must be from organism B.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="danRerLib.mapping._check_column_name_matches_id_choice">
<span class="sig-prename descclassname"><span class="pre">danRerLib.mapping.</span></span><span class="sig-name descname"><span class="pre">_check_column_name_matches_id_choice</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_from</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/danRerLib/mapping.html#_check_column_name_matches_id_choice"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#danRerLib.mapping._check_column_name_matches_id_choice" title="Permalink to this definition"></a></dt>
<dd><p>Check if the specified column name matches the chosen gene ID type.</p>
<p>This function ensures that the specified column name exists in the DataFrame if provided,
or that the chosen gene ID type matches a column name in the DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>pd.DataFrame</em>) – The DataFrame containing gene ID information.</p></li>
<li><p><strong>id_from</strong> (<em>str</em>) – The chosen gene ID type.</p></li>
<li><p><strong>column_name</strong> (<em>str</em>) – The name of the column containing gene IDs (if specified).</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#danRerLib.mapping.InvalidGeneTypeError" title="danRerLib.mapping.InvalidGeneTypeError"><strong>InvalidGeneTypeError</strong></a> – If the specified column name does not exist in the dataset or
if the chosen gene ID type does not match any column name.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>If the ‘column_name’ parameter is provided, this function checks if it exists in the DataFrame.</p></li>
<li><p>If ‘column_name’ is not provided, it checks if ‘id_from’ matches any column name in the DataFrame.</p></li>
<li><p>Ensure that the dataset and column names are accurate to perform gene ID conversion.</p></li>
</ul>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../enrichment/index.html" class="btn btn-neutral float-left" title="danRerLib.enrichment" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../settings/index.html" class="btn btn-neutral float-right" title="danRerLib.settings" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Ashley Schwartz.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>