Search.setIndex({"docnames": ["autoapi/danRerLib/GO/index", "autoapi/danRerLib/KEGG/index", "autoapi/danRerLib/database/index", "autoapi/danRerLib/enrichment/index", "autoapi/danRerLib/index", "autoapi/danRerLib/mapping/index", "autoapi/danRerLib/settings/index", "autoapi/danRerLib/utils/index", "autoapi/index", "changelog", "conduct", "contributing", "index", "tutorials/example", "tutorials/perform_geneid_mapping", "userguide/installation", "userguide/introduction", "userguide/usage"], "filenames": ["autoapi/danRerLib/GO/index.rst", "autoapi/danRerLib/KEGG/index.rst", "autoapi/danRerLib/database/index.rst", "autoapi/danRerLib/enrichment/index.rst", "autoapi/danRerLib/index.rst", "autoapi/danRerLib/mapping/index.rst", "autoapi/danRerLib/settings/index.rst", "autoapi/danRerLib/utils/index.rst", "autoapi/index.rst", "changelog.md", "conduct.md", "contributing.md", "index.md", "tutorials/example.ipynb", "tutorials/perform_geneid_mapping.ipynb", "userguide/installation.md", "userguide/introduction.md", "userguide/usage.md"], "titles": ["<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.GO</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.KEGG</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.database</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.enrichment</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.mapping</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.settings</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.utils</span></code>", "API Reference", "Changelog", "Code of Conduct", "Contributing", "danRerLib Documentation", "Why", "Perform Gene ID Mapping", "Installation", "Introduction", "Usage Principles"], "terms": {"go_ids_path": 0, "go_path_dr": 0, "go_path_drem": 0, "go_path_hsa": 0, "go_basic_url": 0, "http": 0, "purl": 0, "obolibrari": 0, "org": [0, 1, 3], "obo": 0, "basic": 0, "go_zfin_url": 0, "current": [0, 5, 11, 14], "geneontologi": 0, "annot": 0, "zfin": [0, 1, 5, 6, 14], "gaf": 0, "gz": 0, "go_ncbi_url": 0, "ftp": 0, "ncbi": [0, 1, 5, 6, 14], "nlm": 0, "nih": 0, "gov": 0, "gene": [0, 1, 3, 5, 6, 12], "data": [0, 1, 5, 7, 13, 14], "gene2go": 0, "get_genes_in_go_concept": 0, "concept_id": [0, 3], "organ": [0, 1, 5, 14], "gene_id_typ": [0, 3], "none": [0, 1, 3, 5, 7], "desir": [0, 5], "id": [0, 1, 3, 5, 6, 12], "type": [0, 5, 14], "return": [0, 1, 5, 14], "id_exists_given_organ": 0, "check_id_format": 0, "is_numer": 0, "valu": [0, 3, 5], "build_al": 0, "thi": [1, 3, 5, 8, 10, 11, 14, 16], "provid": [1, 5], "retriev": [1, 5, 14], "inform": [1, 5, 10, 14], "associ": 1, "pathwai": [1, 3, 14], "diseas": 1, "variou": 1, "includ": [1, 5, 10, 11, 14], "human": [1, 5, 6], "zebrafish": [1, 5, 14], "map": [1, 4, 8, 12], "get_genes_in_pathwai": 1, "specif": [1, 10, 14], "get_genes_in_diseas": 1, "specifi": [1, 5, 14], "constant": [1, 5], "ncbi_id": [1, 3, 5, 6], "identifi": [1, 5], "zfin_id": [1, 3, 5, 6], "ens_id": [1, 5, 6], "ensembl": [1, 5, 6, 14], "symbol": [1, 5, 6, 14], "human_id": [1, 5, 6], "databas": [1, 3, 5, 12, 14], "rebuild": [1, 5], "build_kegg_databas": 1, "build": [1, 5, 14], "updat": [1, 11], "note": [1, 5, 11, 14], "i": [1, 3, 5, 10, 11, 14], "design": [1, 5], "access": 1, "from": [1, 5, 10], "It": [1, 5, 14], "obtain": 1, "either": [1, 14], "through": [1, 11, 14], "direct": 1, "api": [1, 12], "call": [1, 14], "read": [1, 5, 14], "pre": 1, "process": [1, 5, 14], "file": [1, 5], "option": [1, 5, 14], "dre": [1, 3], "danio": 1, "rerio": 1, "hsa": 1, "drem": 1, "exampl": [1, 5, 10, 12, 14], "To": [1, 5, 14], "pathway_gen": 1, "hsa04010": 1, "For": [1, 5, 14], "detail": [1, 5, 10, 11], "each": [1, 5, 14], "usag": [1, 5, 12], "pleas": [1, 5, 11, 14], "refer": [1, 5, 12, 14], "document": [1, 5, 8, 14], "more": [1, 5, 14], "full": [1, 5, 14], "tutori": [1, 5, 14], "pathway_id": 1, "paramet": [1, 5, 14], "str": [1, 3, 5, 7], "The": [1, 5, 10, 11, 14], "interest": [1, 3, 10], "which": [1, 10, 14], "default": [1, 3, 5, 14], "us": [1, 3, 5, 10, 11, 14], "pd": [1, 3, 5, 7, 13, 14], "datafram": [1, 3, 5, 7, 14], "A": [1, 3, 5, 14], "panda": [1, 5, 7, 13, 14], "contain": [1, 3, 5, 8, 14], "should": [1, 11, 14], "valid": 1, "mai": [1, 10, 14], "requir": [1, 3, 5, 11], "internet": 1, "download": [1, 11], "disease_id": 1, "bool": [1, 5], "fals": [1, 3, 5, 14], "ar": [1, 5, 10, 11, 14], "same": [1, 14], "result": [1, 5], "character": 1, "whether": [1, 5], "rest": 1, "If": [1, 5, 11, 14], "set": [1, 3, 5, 10, 11, 14], "true": [1, 3, 5, 14], "fetch": 1, "intend": 1, "creation": 1, "run": [1, 5, 14], "onli": [1, 14], "dure": 1, "version": [1, 10, 11], "take": [1, 10], "some": [1, 5, 14], "time": [1, 14], "so": 1, "exercis": 1, "caution": 1, "when": [1, 5, 10, 11, 14], "perform": [1, 3, 5, 12], "follow": [1, 5, 10, 14], "step": [1, 11, 14], "store": [1, 14], "them": [1, 5, 14], "creat": [1, 5, 8, 10, 11], "directori": [1, 5, 14], "done": [1, 11, 14], "carefulli": 1, "involv": 1, "logist": 3, "gene_univers": 3, "gene_set": 3, "concept_typ": 3, "concept_nam": 3, "sig_cutoff": 3, "0": [3, 13, 14], "05": 3, "regress": 3, "method": [3, 14], "p": 3, "cutoff": 3, "taken": 3, "fisher": 3, "all": [3, 10, 11, 14], "concept": 3, "kegg": [3, 4, 8], "go": [3, 4, 8, 14], "etc": 3, "e": [3, 10], "g": 3, "enrich_kegg": 3, "sig_gene_cutoff_pvalu": 3, "sig_conceptid_cutoff_pvalu": 3, "sig_conceptid_cutoff_fdr": 3, "order_by_p_valu": 3, "given": [3, 5, 11, 14], "dataset": [3, 5], "pvalu": [3, 14], "log2fc": [3, 14], "order": [3, 5, 14], "enrich_go": 3, "test_kegg_enrich": 3, "option1": 3, "option2": 3, "option3": 3, "investig": [3, 10], "test_go_enrich": 3, "option4": 3, "option5": 3, "test": [3, 11, 12, 14], "enrich": [4, 8], "orthologi": 5, "check": [5, 11, 14], "between": [5, 14], "differ": [5, 10, 14], "speci": 5, "focu": 5, "convert_id": [5, 14], "convert": 5, "list": [5, 7], "add_mapped_column": [5, 14], "add": [5, 11, 14], "new": [5, 14], "column": 5, "convert_to_human": 5, "correspond": 5, "ortholog": 5, "convert_to_zebrafish": 5, "add_mapped_ortholog_column": 5, "build_gene_map": 5, "master": [5, 14], "base": [5, 14], "convers": [5, 14], "build_ortho_map": 5, "genom": [5, 14], "research": 5, "project": [5, 10, 11], "you": [5, 11, 14], "would": [5, 11, 14], "like": [5, 14], "re": [5, 11], "ensur": 5, "your": [5, 11, 14], "accur": 5, "name": [5, 11, 14], "per": 5, "recommend": 5, "packag": [5, 14], "gene_list": 5, "ensdarg00000012345": 5, "ensdarg00000067890": 5, "converted_gen": 5, "except": 5, "databasenotfounderror": 5, "rais": 5, "found": 5, "invalidgenetypeerror": 5, "invalid": 5, "union": 5, "seri": [5, 7], "numpi": 5, "arrai": 5, "id_from": [5, 14], "id_to": [5, 14], "keep_map": [5, 14], "out_format": 5, "support": [5, 11, 14], "format": [5, 11, 14], "np": 5, "keep": [5, 11], "origin": [5, 14], "output": [5, 14], "other": [5, 10, 14], "import": [5, 13, 14], "have": [5, 10, 11, 14], "otherwis": [5, 10, 14], "guarante": 5, "column_name_with_id": [5, 14], "keep_old_id": [5, 14], "drop_na": 5, "ha": [5, 14], "must": [5, 14], "one": [5, 14], "doesn": 5, "t": [5, 14], "match": [5, 14], "old": [5, 14], "drop": [5, 14], "row": [5, 14], "na": 5, "ad": [5, 14], "input": 5, "unless": 5, "zfish_gene_typ": 5, "keep_missing_ortho": 5, "retain": 5, "miss": 5, "target": 5, "get_ortho_id": 5, "can": [5, 11, 14], "master_gene_mapping_file_v": 5, "version_num": [5, 6], "txt": [5, 12, 13, 14], "raw_data": 5, "subdirectori": [5, 14], "zfin_to_ncbi_v": 5, "zfin_to_ensembl_v": 5, "filenotfounderror": 5, "both": [5, 10, 14], "folder": [5, 14], "work": [5, 11, 14], "merg": 5, "zfin_to_ncbi": 5, "zfin_to_ensembl": 5, "remov": [5, 10], "unnecessari": 5, "save": [5, 14], "make": [5, 10, 11], "sure": 5, "befor": [5, 11, 14], "need": [5, 14], "place": 5, "master_ortho_mapping_file_v": 5, "zfish_human_orthology_v": 5, "zfish_human_orthologi": 5, "duplic": 5, "entri": [5, 14], "1": [6, 8, 10, 13, 14], "file_dir": 6, "database_dir": 6, "raw_data_dir": 6, "kegg_data_dir": 6, "go_data_dir": 6, "master_mapping_file_path": 6, "master_ortho_file_path": 6, "pretty_print_seri": [7, 14], "save_data": [7, 14], "locat": [7, 14], "download_url": 7, "url_to_download": 7, "download_url_to_fil": 7, "output_file_path": 7, "page": 8, "auto": 8, "gener": [8, 14], "danrerlib": [8, 9, 11, 14], "sphinx": 8, "autoapi": [8, 14], "first": [9, 14], "releas": [9, 11], "In": [10, 14], "foster": 10, "an": [10, 14, 16], "open": [10, 11], "welcom": [10, 11], "environ": 10, "we": [10, 14], "contributor": 10, "maintain": [10, 14], "particip": 10, "commun": 10, "harass": 10, "free": [10, 11, 14], "experi": 10, "everyon": 10, "regardless": 10, "ag": 10, "bodi": 10, "size": 10, "disabl": 10, "ethnic": 10, "gender": 10, "ident": 10, "express": 10, "level": 10, "nation": 10, "person": 10, "appear": 10, "race": 10, "religion": 10, "sexual": 10, "orient": 10, "behavior": 10, "contribut": [10, 12], "posit": 10, "inclus": 10, "languag": 10, "Being": 10, "respect": 10, "viewpoint": 10, "gracefulli": 10, "accept": 10, "construct": 10, "critic": 10, "focus": 10, "what": [10, 14], "best": [10, 14], "show": 10, "empathi": 10, "toward": 10, "member": 10, "unaccept": 10, "imageri": 10, "unwelcom": 10, "attent": 10, "advanc": 10, "troll": 10, "insult": 10, "derogatori": 10, "comment": 10, "polit": 10, "attack": 10, "public": 10, "privat": 10, "publish": 10, "physic": 10, "electron": 10, "address": 10, "without": [10, 14], "explicit": 10, "permiss": 10, "could": [10, 14], "reason": 10, "consid": 10, "inappropri": 10, "profession": 10, "clarifi": 10, "expect": 10, "appropri": [10, 11], "fair": 10, "correct": 10, "action": 10, "ani": [10, 11, 14], "instanc": 10, "right": 10, "edit": 10, "reject": 10, "commit": [10, 11], "wiki": 10, "issu": [10, 11, 14], "align": 10, "ban": 10, "temporarili": 10, "perman": 10, "thei": [10, 11, 14], "deem": 10, "threaten": 10, "offens": 10, "harm": 10, "appli": 10, "within": 10, "space": 10, "individu": 10, "repres": 10, "its": [10, 11], "offici": [10, 11], "mail": 10, "post": [10, 11], "via": 10, "social": 10, "media": 10, "account": 10, "act": 10, "appoint": 10, "onlin": 10, "offlin": 10, "event": 10, "represent": 10, "further": 10, "defin": [10, 14], "abus": 10, "report": 10, "contact": 10, "team": 10, "review": 10, "complaint": 10, "respond": 10, "wai": [10, 14], "circumst": 10, "oblig": 10, "confidenti": 10, "regard": 10, "incid": 10, "polici": 10, "separ": [10, 14], "who": 10, "do": [10, 14], "good": 10, "faith": 10, "face": 10, "temporari": 10, "repercuss": 10, "determin": 10, "": [10, 11], "leadership": 10, "adapt": 10, "coven": 10, "homepag": 10, "4": [10, 13, 14], "greatli": 11, "appreci": 11, "everi": 11, "littl": 11, "bit": 11, "help": 11, "credit": 11, "alwai": [11, 14], "oper": 11, "system": 11, "about": [11, 14], "local": 11, "setup": [11, 12], "might": [11, 14], "troubleshoot": 11, "reproduc": 11, "look": [11, 14], "github": 11, "anyth": 11, "tag": 11, "want": [11, 14], "whoever": 11, "enhanc": 11, "never": 11, "enough": 11, "feel": [11, 14], "part": 11, "doc": [11, 12, 14], "docstr": 11, "even": [11, 14], "web": 11, "blog": 11, "articl": 11, "propos": 11, "explain": 11, "how": [11, 14], "scope": 11, "narrow": 11, "possibl": 11, "easier": 11, "rememb": [11, 14], "volunt": 11, "driven": 11, "readi": 11, "here": [11, 14], "up": 11, "develop": 11, "copi": [11, 14], "instal": [11, 12, 14], "poetri": 11, "git": 11, "similar": 11, "branch": 11, "chang": [11, 14], "checkout": 11, "b": [11, 14], "bugfix": 11, "conform": 11, "pass": 11, "meet": 11, "addit": 11, "function": [11, 14], "python": [11, 14], "By": [11, 14], "agre": 11, "abid": 11, "term": 11, "readm": 12, "md": 12, "py": 12, "licens": 12, "src": 12, "__init__": 12, "datafil": 12, "introduct": 12, "principl": 12, "why": 12, "changelog": 12, "code": [12, 14], "conduct": 12, "2": [13, 14], "read_csv": [13, 14], "trial": 13, "3": [13, 14], "notebook": [], "cell": [], "print": 14, "ipython": [], "core": [], "interactiveshel": [], "ast_node_interact": [], "author": 14, "ashlei": 14, "schwartz": 14, "date": 14, "june": 14, "30": 14, "2023": 14, "goe": 14, "over": 14, "simpli": 14, "anoth": 14, "come": 14, "veri": 14, "form": 14, "depend": 14, "get": 14, "confus": 14, "descript": 14, "start": 14, "zdb": 14, "zebafish": 14, "011219": 14, "link": 14, "integ": 14, "manag": 14, "also": 14, "known": 14, "entrez": 14, "140634": 14, "refseq": 14, "cyp1a": 14, "nomenclatur": 14, "ensdar": 14, "ensdarg00000098315": 14, "util": 14, "two": 14, "kei": 14, "element": 14, "sampl": 14, "csv": 14, "tsv": 14, "past": 14, "test_data": 14, "rel": 14, "path": 14, "example_diff_express_data": 14, "while": 14, "larg": 14, "foundat": 14, "execut": [], "understand": 14, "absolut": 14, "spell": 14, "case": [], "sensit": 14, "sy": [], "pathlib": [], "intro": 16, "relev": 14, "modul": 14, "see": 14, "There": 14, "varieti": 14, "scenario": 14, "most": 14, "simplest": 14, "few": 14, "sinc": 14, "common": 14, "entir": 14, "excel": 14, "fil": 14, "small": 14, "especi": 14, "just": 14, "list_of_gene_id": 14, "100000252": 14, "100000750": 14, "100001198": 14, "100001260": 14, "100002225": 14, "100002263": 14, "100002756": 14, "100003223": 14, "100007521": 14, "100149273": 14, "100149794": 14, "100170795": 14, "100321746": 14, "100329897": 14, "100330617": 14, "tell": 14, "program": 14, "begin": 14, "don": 14, "worri": 14, "let": 14, "know": 14, "made": 14, "mistak": 14, "launch": 14, "current_gene_id_typ": 14, "desired_gene_id_typ": 14, "mean": 14, "our": 14, "librari": 14, "onc": 14, "converted_id": 14, "variabl": 14, "visual": 14, "shell": 14, "fine": 14, "idea": 14, "print_series_pretti": 14, "030131": 14, "1904": 14, "3404": 14, "030325": 14, "030616": 14, "609": 14, "040426": 14, "743": 14, "050309": 14, "246": 14, "071009": 14, "6": 14, "080219": 14, "34": 14, "080723": 14, "44": 14, "081223": 14, "090313": 14, "141": 14, "091117": 14, "28": 14, "110309": 14, "120215": 14, "92": 14, "rather": 14, "previous": 14, "choic": 14, "file_nam": 14, "out_data_dir": [], "nameerror": [], "traceback": [], "recent": [], "last": [], "line": 14, "out_data": 14, "easili": 14, "repeat": 14, "out": 14, "data_file_path": 14, "ncbi_gene_list": 14, "sep": 14, "7": 14, "000616": 14, "10": 14, "1024": [], "174": [], "8872": [], "030219": [], "83": [], "040718": [], "198": [], "040808": [], "040912": [], "16": [], "050419": 14, "85": 14, "050522": [], "533": [], "080415": 14, "081105": [], "136": [], "150409": [], "991229": 14, "12": 14, "manner": 14, "As": 14, "now": 14, "below": 14, "limit": 14, "non": 14, "quit": 14, "than": 14, "sometim": 14, "along": 14, "next": 14, "instruct": 14, "opt": [], "anaconda3": [], "env": [], "bioinfo": [], "lib": [], "python3": [], "9": 14, "site": [], "io": [], "parser": [], "reader": [], "948": [], "filepath_or_buff": [], "delimit": [], "header": [], "index_col": [], "usecol": [], "dtype": [], "engin": [], "true_valu": [], "false_valu": [], "skipinitialspac": [], "skiprow": [], "skipfoot": [], "nrow": [], "na_valu": [], "keep_default_na": [], "na_filt": [], "verbos": [], "skip_blank_lin": [], "parse_d": [], "infer_datetime_format": [], "keep_date_col": [], "date_pars": [], "date_format": [], "dayfirst": [], "cache_d": [], "iter": [], "chunksiz": [], "compress": [], "thousand": [], "decim": [], "linetermin": [], "quotechar": [], "quot": [], "doublequot": [], "escapechar": [], "encod": [], "encoding_error": [], "dialect": [], "on_bad_lin": [], "delim_whitespac": [], "low_memori": [], "memory_map": [], "float_precis": [], "storage_opt": [], "dtype_backend": [], "935": [], "kwds_default": [], "_refine_defaults_read": [], "936": [], "937": [], "944": [], "945": [], "946": [], "kwd": [], "_read": [], "611": [], "608": [], "_validate_nam": [], "610": [], "textfileread": [], "613": [], "614": [], "1448": [], "self": [], "f": 14, "1445": [], "has_index_nam": [], "1447": [], "handl": [], "iohandl": [], "_engin": [], "_make_engin": [], "1705": [], "1703": [], "mode": [], "1704": [], "get_handl": [], "1706": [], "1707": [], "1708": [], "1709": [], "1710": [], "1711": [], "is_text": [], "1712": [], "error": [], "strict": [], "1713": [], "1714": [], "1715": [], "assert": [], "1716": [], "863": [], "path_or_buf": [], "858": [], "elif": [], "isinst": [], "859": [], "filenam": [], "binari": [], "860": [], "doe": 14, "newlin": [], "861": [], "ioarg": [], "862": [], "864": [], "865": [], "866": [], "867": [], "868": [], "869": [], "870": [], "els": [], "871": [], "872": [], "errno": [], "No": [], "tabl": 14, "abov": 14, "am": 14, "my": 14, "converst": 14, "activ": 14, "earlier": 14, "converted_id_t": 14, "5": 14, "8": 14, "11": 14, "13": 14, "head": 14, "automat": 14, "besid": 14, "initi": 14, "57924": 14, "100003844": [], "337772": [], "336928": [], "338183": [], "436767": [], "445326": [], "447813": [], "794572": 14, "553785": [], "564772": 14, "100003419": [], "105665603": [], "30459": 14, "methodologi": 14, "great": 14, "where": 14, "properli": 14, "sort": 14, "extrem": 14, "30301": 14, "14": [], "x": 14, "y": 14, "z": 14, "logfc": 14, "03": 14, "essenti": 14, "geneid": 14, "often": 14, "deal": 14, "lot": 14, "sub": 14, "extens": 14, "realli": 14, "quick": 14, "three": 14, "stat": 14, "col": 14, "shape": 14, "100000006": 14, "792615": 14, "115009": 14, "100000044": 14, "015286": 14, "803879": 14, "100000085": 14, "264762": 14, "267360": 14, "5464": 14, "give": 14, "new_data": 14, "5654": 14, "121214": 14, "60": 14, "1312": 14, "converted_datafram": 14, "remind": 14, "test_set_invalid_col_nam": 14, "100000009": 14, "607285": 14, "144714": 14, "100000026": 14, "021338": 14, "603871": 14, "around": 14, "exactli": 14, "Of": 14, "cours": 14, "convent": 14, "seen": 14, "130530": 14, "778": 14, "120823": 14, "wa": 14, "successfulli": 14, "though": 14, "final": 14, "anymor": 14, "therefor": 14, "longer": 14, "conclud": 14, "summari": 14, "exist": 14}, "objects": {"": [[4, 0, 0, "-", "danRerLib"]], "danRerLib": [[0, 0, 0, "-", "GO"], [1, 0, 0, "-", "KEGG"], [2, 0, 0, "-", "database"], [3, 0, 0, "-", "enrichment"], [5, 0, 0, "-", "mapping"], [6, 0, 0, "-", "settings"], [7, 0, 0, "-", "utils"]], "danRerLib.GO": [[0, 1, 1, "", "GO_BASIC_URL"], [0, 1, 1, "", "GO_IDS_PATH"], [0, 1, 1, "", "GO_NCBI_URL"], [0, 1, 1, "", "GO_PATH_dre"], [0, 1, 1, "", "GO_PATH_dreM"], [0, 1, 1, "", "GO_PATH_hsa"], [0, 1, 1, "", "GO_ZFIN_URL"], [0, 2, 1, "", "build_all"], [0, 2, 1, "", "check_id_format"], [0, 2, 1, "", "get_genes_in_GO_concept"], [0, 2, 1, "", "id_exists_given_organism"], [0, 2, 1, "", "is_numeric"]], "danRerLib.KEGG": [[1, 2, 1, "", "build_kegg_database"], [1, 2, 1, "", "get_genes_in_disease"], [1, 2, 1, "", "get_genes_in_pathway"]], "danRerLib.enrichment": [[3, 2, 1, "", "enrich_GO"], [3, 2, 1, "", "enrich_KEGG"], [3, 2, 1, "", "fishers"], [3, 2, 1, "", "investigation"], [3, 2, 1, "", "logistic"], [3, 2, 1, "", "test_GO_enrich"], [3, 2, 1, "", "test_KEGG_enrich"], [3, 2, 1, "", "testing"]], "danRerLib.mapping": [[5, 3, 1, "", "DatabaseNotFoundError"], [5, 3, 1, "", "InvalidGeneTypeError"], [5, 2, 1, "", "add_mapped_column"], [5, 2, 1, "", "add_mapped_ortholog_column"], [5, 2, 1, "", "build_gene_mapping"], [5, 2, 1, "", "build_ortho_mapping"], [5, 2, 1, "", "convert_ids"], [5, 2, 1, "", "convert_to_human"], [5, 2, 1, "", "convert_to_zebrafish"], [5, 2, 1, "", "get_ortho_ids"]], "danRerLib.settings": [[6, 1, 1, "", "DATABASE_DIR"], [6, 1, 1, "", "ENS_ID"], [6, 1, 1, "", "FILE_DIR"], [6, 1, 1, "", "GO_DATA_DIR"], [6, 1, 1, "", "HUMAN_ID"], [6, 1, 1, "", "KEGG_DATA_DIR"], [6, 1, 1, "", "MASTER_MAPPING_FILE_PATH"], [6, 1, 1, "", "MASTER_ORTHO_FILE_PATH"], [6, 1, 1, "", "NCBI_ID"], [6, 1, 1, "", "RAW_DATA_DIR"], [6, 1, 1, "", "SYMBOL"], [6, 1, 1, "", "VERSION_NUM"], [6, 1, 1, "", "ZFIN_ID"]], "danRerLib.utils": [[7, 2, 1, "", "download_url"], [7, 2, 1, "", "download_url_to_file"], [7, 2, 1, "", "pretty_print_series"], [7, 2, 1, "", "save_data"]]}, "objtypes": {"0": "py:module", "1": "py:data", "2": "py:function", "3": "py:exception"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "data", "Python data"], "2": ["py", "function", "Python function"], "3": ["py", "exception", "Python exception"]}, "titleterms": {"danrerlib": [0, 1, 2, 3, 4, 5, 6, 7, 12, 16], "go": 0, "modul": [0, 1, 3, 5, 6, 7], "content": [0, 1, 3, 5, 6, 7], "function": [0, 1, 3, 5, 7], "attribut": [0, 10], "kegg": 1, "databas": [2, 17], "enrich": [3, 17], "submodul": 4, "map": [5, 14], "set": 6, "util": 7, "api": 8, "refer": 8, "changelog": 9, "v0": 9, "1": 9, "0": 9, "21": 9, "09": 9, "2023": 9, "code": [10, 11], "conduct": [10, 11], "our": 10, "pledg": 10, "standard": 10, "respons": 10, "scope": 10, "enforc": 10, "contribut": 11, "type": [11, 17], "report": 11, "bug": 11, "fix": 11, "implement": 11, "featur": 11, "write": 11, "document": [11, 12], "submit": 11, "feedback": 11, "get": 11, "start": 11, "pull": 11, "request": 11, "guidelin": 11, "user": 12, "": 12, "guid": 12, "tutori": 12, "addit": 12, "inform": 12, "why": 13, "perform": 14, "gene": [14, 17], "id": [14, 17], "purpos": 14, "background": 14, "requir": 14, "instal": 15, "anaconda": 15, "pypi": 15, "develop": 15, "version": 15, "introduct": 16, "what": 16, "i": 16, "some": 16, "other": 16, "thing": 16, "usag": 17, "principl": 17, "workflow": 17, "pathwai": 17, "execut": 14, "simpl": 14, "case": 14, "convert": 14, "list": 14, "from": 14, "file": 14, "keep": 14, "column": 14, "larger": 14, "dataset": 14, "special": 14}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx": 57}, "alltitles": {"Changelog": [[9, "changelog"]], "v0.1.0 (21/09/2023)": [[9, "v0-1-0-21-09-2023"]], "Code of Conduct": [[10, "code-of-conduct"], [11, "code-of-conduct"]], "Our Pledge": [[10, "our-pledge"]], "Our Standards": [[10, "our-standards"]], "Our Responsibilities": [[10, "our-responsibilities"]], "Scope": [[10, "scope"]], "Enforcement": [[10, "enforcement"]], "Attribution": [[10, "attribution"]], "Contributing": [[11, "contributing"]], "Types of Contributions": [[11, "types-of-contributions"]], "Report Bugs": [[11, "report-bugs"]], "Fix Bugs": [[11, "fix-bugs"]], "Implement Features": [[11, "implement-features"]], "Write Documentation": [[11, "write-documentation"]], "Submit Feedback": [[11, "submit-feedback"]], "Get Started!": [[11, "get-started"]], "Pull Request Guidelines": [[11, "pull-request-guidelines"]], "danRerLib Documentation": [[12, "danrerlib-documentation"]], "User's Guide": [[12, null]], "Tutorials": [[12, null]], "Additional Information": [[12, null]], "Installation": [[15, "installation"]], "Anaconda": [[15, "anaconda"]], "PyPi": [[15, "pypi"]], "Development Version": [[15, "development-version"]], "Introduction": [[16, "introduction"]], "What is danRerLib?": [[16, "what-is-danrerlib"]], "some other things\u2026": [[16, "some-other-things"]], "Usage Principles": [[17, "usage-principles"]], "Workflow": [[17, "workflow"]], "Gene ID Types": [[17, "gene-id-types"]], "Pathway Databases": [[17, "pathway-databases"]], "Enrichment Types": [[17, "enrichment-types"]], "Why": [[13, "why"]], "danRerLib.GO": [[0, "module-danRerLib.GO"]], "Module Contents": [[0, "module-contents"], [1, "module-contents"], [3, "module-contents"], [5, "module-contents"], [6, "module-contents"], [7, "module-contents"]], "Functions": [[0, "functions"], [1, "functions"], [3, "functions"], [5, "functions"], [7, "functions"]], "Attributes": [[0, "attributes"]], "danRerLib.KEGG": [[1, "module-danRerLib.KEGG"]], "danRerLib.database": [[2, "module-danRerLib.database"]], "danRerLib.enrichment": [[3, "module-danRerLib.enrichment"]], "danRerLib": [[4, "module-danRerLib"]], "Submodules": [[4, "submodules"]], "danRerLib.mapping": [[5, "module-danRerLib.mapping"]], "danRerLib.settings": [[6, "module-danRerLib.settings"]], "danRerLib.utils": [[7, "module-danRerLib.utils"]], "API Reference": [[8, "api-reference"]], "Perform Gene ID Mapping": [[14, "perform-gene-id-mapping"]], "Purpose and Background": [[14, "purpose-and-background"]], "Requirements": [[14, "requirements"]], "Execute Mappings": [[14, "execute-mappings"]], "Simple Case: Convert a list of Gene IDs": [[14, "simple-case-convert-a-list-of-gene-ids"]], "Simple Case: Convert a list of Gene IDs From a File": [[14, "simple-case-convert-a-list-of-gene-ids-from-a-file"]], "Simple Case: Convert a list of Gene IDs and Keep Mapping": [[14, "simple-case-convert-a-list-of-gene-ids-and-keep-mapping"]], "Convert Gene IDs in a Column of a Larger Dataset": [[14, "convert-gene-ids-in-a-column-of-a-larger-dataset"]], "Special Cases": [[14, "special-cases"]]}, "indexentries": {"go_basic_url (in module danrerlib.go)": [[0, "danRerLib.GO.GO_BASIC_URL"]], "go_ids_path (in module danrerlib.go)": [[0, "danRerLib.GO.GO_IDS_PATH"]], "go_ncbi_url (in module danrerlib.go)": [[0, "danRerLib.GO.GO_NCBI_URL"]], "go_path_dre (in module danrerlib.go)": [[0, "danRerLib.GO.GO_PATH_dre"]], "go_path_drem (in module danrerlib.go)": [[0, "danRerLib.GO.GO_PATH_dreM"]], "go_path_hsa (in module danrerlib.go)": [[0, "danRerLib.GO.GO_PATH_hsa"]], "go_zfin_url (in module danrerlib.go)": [[0, "danRerLib.GO.GO_ZFIN_URL"]], "build_all() (in module danrerlib.go)": [[0, "danRerLib.GO.build_all"]], "check_id_format() (in module danrerlib.go)": [[0, "danRerLib.GO.check_id_format"]], "danrerlib.go": [[0, "module-danRerLib.GO"]], "get_genes_in_go_concept() (in module danrerlib.go)": [[0, "danRerLib.GO.get_genes_in_GO_concept"]], "id_exists_given_organism() (in module danrerlib.go)": [[0, "danRerLib.GO.id_exists_given_organism"]], "is_numeric() (in module danrerlib.go)": [[0, "danRerLib.GO.is_numeric"]], "module": [[0, "module-danRerLib.GO"], [1, "module-danRerLib.KEGG"], [2, "module-danRerLib.database"], [3, "module-danRerLib.enrichment"], [4, "module-danRerLib"], [5, "module-danRerLib.mapping"], [6, "module-danRerLib.settings"], [7, "module-danRerLib.utils"]], "build_kegg_database() (in module danrerlib.kegg)": [[1, "danRerLib.KEGG.build_kegg_database"]], "danrerlib.kegg": [[1, "module-danRerLib.KEGG"]], "get_genes_in_disease() (in module danrerlib.kegg)": [[1, "danRerLib.KEGG.get_genes_in_disease"]], "get_genes_in_pathway() (in module danrerlib.kegg)": [[1, "danRerLib.KEGG.get_genes_in_pathway"]], "danrerlib.database": [[2, "module-danRerLib.database"]], "danrerlib.enrichment": [[3, "module-danRerLib.enrichment"]], "enrich_go() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.enrich_GO"]], "enrich_kegg() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.enrich_KEGG"]], "fishers() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.fishers"]], "investigation() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.investigation"]], "logistic() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.logistic"]], "test_go_enrich() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.test_GO_enrich"]], "test_kegg_enrich() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.test_KEGG_enrich"]], "testing() (in module danrerlib.enrichment)": [[3, "danRerLib.enrichment.testing"]], "danrerlib": [[4, "module-danRerLib"]], "databasenotfounderror": [[5, "danRerLib.mapping.DatabaseNotFoundError"]], "invalidgenetypeerror": [[5, "danRerLib.mapping.InvalidGeneTypeError"]], "add_mapped_column() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.add_mapped_column"]], "add_mapped_ortholog_column() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.add_mapped_ortholog_column"]], "build_gene_mapping() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.build_gene_mapping"]], "build_ortho_mapping() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.build_ortho_mapping"]], "convert_ids() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.convert_ids"]], "convert_to_human() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.convert_to_human"]], "convert_to_zebrafish() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.convert_to_zebrafish"]], "danrerlib.mapping": [[5, "module-danRerLib.mapping"]], "get_ortho_ids() (in module danrerlib.mapping)": [[5, "danRerLib.mapping.get_ortho_ids"]], "database_dir (in module danrerlib.settings)": [[6, "danRerLib.settings.DATABASE_DIR"]], "ens_id (in module danrerlib.settings)": [[6, "danRerLib.settings.ENS_ID"]], "file_dir (in module danrerlib.settings)": [[6, "danRerLib.settings.FILE_DIR"]], "go_data_dir (in module danrerlib.settings)": [[6, "danRerLib.settings.GO_DATA_DIR"]], "human_id (in module danrerlib.settings)": [[6, "danRerLib.settings.HUMAN_ID"]], "kegg_data_dir (in module danrerlib.settings)": [[6, "danRerLib.settings.KEGG_DATA_DIR"]], "master_mapping_file_path (in module danrerlib.settings)": [[6, "danRerLib.settings.MASTER_MAPPING_FILE_PATH"]], "master_ortho_file_path (in module danrerlib.settings)": [[6, "danRerLib.settings.MASTER_ORTHO_FILE_PATH"]], "ncbi_id (in module danrerlib.settings)": [[6, "danRerLib.settings.NCBI_ID"]], "raw_data_dir (in module danrerlib.settings)": [[6, "danRerLib.settings.RAW_DATA_DIR"]], "symbol (in module danrerlib.settings)": [[6, "danRerLib.settings.SYMBOL"]], "version_num (in module danrerlib.settings)": [[6, "danRerLib.settings.VERSION_NUM"]], "zfin_id (in module danrerlib.settings)": [[6, "danRerLib.settings.ZFIN_ID"]], "danrerlib.settings": [[6, "module-danRerLib.settings"]], "danrerlib.utils": [[7, "module-danRerLib.utils"]], "download_url() (in module danrerlib.utils)": [[7, "danRerLib.utils.download_url"]], "download_url_to_file() (in module danrerlib.utils)": [[7, "danRerLib.utils.download_url_to_file"]], "pretty_print_series() (in module danrerlib.utils)": [[7, "danRerLib.utils.pretty_print_series"]], "save_data() (in module danrerlib.utils)": [[7, "danRerLib.utils.save_data"]]}})