Search.setIndex({"docnames": [".pytest_cache/README", "autoapi/danRerLib/GO/index", "autoapi/danRerLib/KEGG/index", "autoapi/danRerLib/database/index", "autoapi/danRerLib/enrichment/index", "autoapi/danRerLib/index", "autoapi/danRerLib/mapping/index", "autoapi/danRerLib/settings/index", "autoapi/danRerLib/utils/index", "autoapi/index", "changelog", "conduct", "contributing", "example", "index"], "filenames": [".pytest_cache/README.md", "autoapi/danRerLib/GO/index.rst", "autoapi/danRerLib/KEGG/index.rst", "autoapi/danRerLib/database/index.rst", "autoapi/danRerLib/enrichment/index.rst", "autoapi/danRerLib/index.rst", "autoapi/danRerLib/mapping/index.rst", "autoapi/danRerLib/settings/index.rst", "autoapi/danRerLib/utils/index.rst", "autoapi/index.rst", "changelog.md", "conduct.md", "contributing.md", "example.ipynb", "index.md"], "titles": ["pytest cache directory", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.GO</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.KEGG</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.database</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.enrichment</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.mapping</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.settings</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danRerLib.utils</span></code>", "API Reference", "Changelog", "Code of Conduct", "Contributing", "Example usage", "&lt;no title&gt;"], "terms": {"go_ids_path": 1, "go_path_dr": 1, "go_path_drem": 1, "go_path_hsa": 1, "go_basic_url": 1, "http": 1, "purl": 1, "obolibrari": 1, "org": [1, 2, 4], "obo": 1, "basic": 1, "go_zfin_url": 1, "current": [1, 6, 12], "geneontologi": 1, "annot": 1, "zfin": [1, 6, 7], "gaf": 1, "gz": 1, "go_ncbi_url": 1, "ftp": 1, "ncbi": [1, 2, 6, 7], "nlm": 1, "nih": 1, "gov": 1, "gene": [1, 2, 4, 6, 7], "data": [0, 1, 2, 6, 8], "gene2go": 1, "get_genes_in_go_concept": 1, "concept_id": [1, 4], "organ": [1, 2, 4, 6], "gene_id_typ": [1, 4, 6], "none": [1, 2, 4, 6, 8], "desir": [1, 6], "id": [1, 2, 4, 6, 7], "type": [1, 2, 6], "return": [1, 2, 6], "id_exists_given_organ": 1, "check_id_format": 1, "is_numer": 1, "valu": [1, 4, 6], "build_al": 1, "_build_go_id": 1, "thi": [0, 1, 2, 4, 6, 9, 11, 12], "gener": [1, 9], "list": [1, 2, 4, 6, 8], "all": [1, 2, 4, 11, 12], "name": [1, 2, 6, 12], "ontologi": 1, "belong": 1, "i": [1, 2, 4, 6, 11, 12], "download": [1, 2, 12], "directli": 1, "from": [0, 1, 2, 6, 11], "websit": 1, "us": [1, 2, 4, 6, 11, 12, 13], "api": [1, 2], "_add_to_go_id": 1, "purpos": 1, "add": [1, 6, 12], "two": 1, "column": [1, 2, 6], "identifi": [1, 2, 6], "concept": [1, 4], "exist": [1, 2, 6], "zebrafish": [1, 2, 6], "himan": 1, "_build_zebrafish_go": 1, "_build_human_go": 1, "_build_dre_mapped_go": 1, "_get_ontologi": 1, "namespac": 1, "_process_gobasic_cont": 1, "paramet": [2, 6], "_write_gobasic_to_fil": 1, "go_term": 1, "output_fil": 1, "str": [1, 2, 4, 6, 8], "zebrafish_pathways_path": 2, "human_pathways_path": 2, "mapped_zebrafish_pathways_path": 2, "zebrafish_pathways_dir": 2, "human_pathways_dir": 2, "dre_mapped_dir": 2, "human_disease_path": 2, "get_genes_in_pathwai": 2, "pathway_id": 2, "sourc": [1, 2, 4, 6, 8], "option": [0, 2, 6], "dre": [2, 4], "true": [2, 4, 6], "danio": 2, "rerio": 2, "pathwai": [2, 4], "hsa": 2, "human": [2, 6, 7], "drem": 2, "map": [2, 5, 9], "get_genes_in_diseas": 2, "disease_id": 2, "diseas": 2, "note": [2, 6, 12], "produc": 2, "same": [2, 6], "result": [2, 6], "character": 2, "There": 2, "ar": [2, 6, 11, 12], "so": 2, "mani": 2, "opt": 2, "just": 2, "instead": 2, "save": [2, 6], "each": [2, 6], "file": [2, 6], "_check_for_organ": 2, "_check_if_pathway_id_exist": 2, "build_kegg_databas": 2, "A": [2, 4, 6], "databas": [2, 4, 5, 6, 9, 14], "build": [2, 6], "onli": 2, "run": [2, 6], "ON": [], "version": [0, 2, 11, 12], "updat": [2, 12], "take": [2, 11], "while": [], "care": [], "_download_zebrafish_pathway_id": 2, "_download_human_pathway_id": 2, "_create_mapped_zebrafish_pathway_id": 2, "_download_disease_id": 2, "_download_human_pathway_gen": 2, "_download_zebrafish_pathway_gen": 2, "_build_dre_map": 2, "test": [2, 4, 12, 14], "logist": 4, "gene_univers": 4, "gene_set": 4, "concept_typ": 4, "concept_nam": 4, "sig_cutoff": 4, "0": [4, 13], "05": 4, "perform": [2, 4, 6], "regress": 4, "method": 4, "p": 4, "cutoff": 4, "requir": [2, 4, 6, 12], "default": [2, 4, 6], "taken": 4, "set": [4, 5, 6, 9, 11, 12], "pd": [2, 4, 6, 8], "datafram": [2, 4, 6, 8], "fisher": 4, "interest": [2, 4, 11], "kegg": [4, 5, 9], "go": [4, 5, 9], "etc": 4, "e": [2, 4, 11], "g": [2, 4], "enrich_kegg": 4, "ncbi_id": [4, 6, 7], "sig_gene_cutoff_pvalu": 4, "sig_conceptid_cutoff_pvalu": 4, "sig_conceptid_cutoff_fdr": 4, "order_by_p_valu": 4, "given": [2, 4, 6, 12], "dataset": [4, 6], "contain": [0, 2, 4, 6, 9], "pvalu": 4, "log2fc": 4, "order": [4, 6], "enrich_go": 4, "zfin_id": [4, 6, 7], "_check_gene_univers": 4, "_check_concept_ids_kegg": 4, "_check_concept_ids_go": 4, "_id_exists_given_organ": 4, "df": 4, "_check_go_id_format": 4, "_is_numer": 4, "_check_valid_org": 4, "test_kegg_enrich": 4, "option1": 4, "fals": [2, 4, 6], "option2": 4, "option3": 4, "investig": [4, 11], "test_go_enrich": 4, "option4": 4, "option5": 4, "enrich": [5, 9], "util": [5, 9], "__version__": [5, 13], "script": [], "tool": [], "necessari": 6, "vari": [], "identif": [], "The": [2, 6, 11, 12], "ha": 6, "been": [], "check": [2, 6, 12], "out": [], "readm": 14, "more": [0, 6], "inform": [0, 2, 6, 11], "except": 6, "databasenotfounderror": 6, "base": [2, 6], "rais": [2, 6], "when": [2, 6, 11, 12], "directori": [2, 6], "found": 6, "invalidgenetypeerror": 6, "convert_id": 6, "gene_list": 6, "id_from": 6, "id_to": 6, "keep_map": 6, "out_format": 6, "convert": 6, "param": [], "support": [6, 12], "format": [6, 12], "np": 6, "arrai": 6, "seri": [6, 8], "other": [6, 11], "symbol": [6, 7], "ensembl": [6, 7], "ani": [6, 11, 12], "add_mapped_column": 6, "column_name_with_id": 6, "keep_old_id": 6, "drop_na": 6, "panda": [2, 6, 8], "some": [2, 6], "must": [2, 6], "doe": [2, 6], "match": [2, 6], "you": [6, 12], "would": [6, 12], "like": 6, "keep": [6, 12], "old": 6, "drop": 6, "row": 6, "have": [6, 11, 12], "na": 6, "convert_to_human": 6, "zfish_gene_typ": 6, "keep_missing_ortho": 6, "convert_to_zebrafish": 6, "get_ortho_id": 6, "add_mapped_ortholog_column": 6, "build_gene_map": 6, "master": 6, "master_gene_mapping_file_v": 6, "version_num": [6, 7], "txt": [2, 6, 14], "should": [2, 12], "raw_data": 6, "sub": [], "depend": [], "zfin_to_ncbi_v": 6, "zfin_to_ensembl_v": 6, "build_ortho_map": 6, "orthologi": 6, "master_ortho_mapping_file_v": 6, "convert_to_str": [], "_make_sure_is_pandas_seri": 6, "_check_valid_zebrafish_gene_id_typ": 6, "_check_valid_gene_id_type_for_orthologi": 6, "gene_id_from": 6, "gene_id_to": 6, "_check_column_name_matches_id_choic": 6, "column_nam": 6, "zdb": [], "030131": [], "2931": [], "081031": [], "55": [], "61": [], "110420": [], "1": [7, 9, 11, 13], "ens_id": [6, 7], "human_id": [6, 7], "file_dir": 7, "database_dir": 7, "raw_data_dir": 7, "kegg_data_dir": 7, "go_data_dir": 7, "master_mapping_file_path": 7, "master_ortho_file_path": 7, "pretty_print_seri": 8, "save_data": 8, "locat": 8, "download_url": 8, "url_to_download": 8, "download_url_to_fil": 8, "output_file_path": 8, "page": 9, "auto": 9, "document": [6, 9], "danrerlib": [9, 10, 12, 13, 14], "creat": [2, 6, 9, 11, 12], "sphinx": 9, "autoapi": 9, "first": 10, "releas": [10, 12], "In": 11, "foster": 11, "an": [2, 11], "open": [11, 12], "welcom": [11, 12], "environ": 11, "we": 11, "contributor": 11, "maintain": 11, "make": [6, 11, 12], "particip": 11, "project": [6, 11, 12, 13], "commun": 11, "harass": 11, "free": [11, 12], "experi": 11, "everyon": 11, "regardless": 11, "ag": 11, "bodi": 11, "size": 11, "disabl": 11, "ethnic": 11, "gender": 11, "ident": 11, "express": 11, "level": 11, "nation": 11, "person": 11, "appear": 11, "race": 11, "religion": 11, "sexual": 11, "orient": 11, "exampl": [6, 11, 14], "behavior": 11, "contribut": 11, "posit": 11, "includ": [2, 6, 11, 12], "inclus": 11, "languag": 11, "Being": 11, "respect": 11, "differ": [6, 11], "viewpoint": 11, "gracefulli": 11, "accept": 11, "construct": 11, "critic": 11, "focus": 11, "what": 11, "best": 11, "show": 11, "empathi": 11, "toward": 11, "member": 11, "unaccept": 11, "imageri": 11, "unwelcom": 11, "attent": 11, "advanc": 11, "troll": 11, "insult": 11, "derogatori": 11, "comment": 11, "polit": 11, "attack": 11, "public": 11, "privat": 11, "publish": 11, "physic": 11, "electron": 11, "address": 11, "without": 11, "explicit": 11, "permiss": 11, "which": [0, 2, 6, 11], "could": 11, "reason": 11, "consid": 11, "inappropri": 11, "profession": 11, "clarifi": 11, "expect": 11, "appropri": [11, 12], "fair": 11, "correct": 11, "action": 11, "instanc": 11, "right": 11, "remov": [6, 11], "edit": 11, "reject": 11, "commit": [0, 11, 12], "wiki": 11, "issu": [11, 12], "align": 11, "ban": 11, "temporarili": 11, "perman": 11, "thei": [11, 12], "deem": 11, "threaten": 11, "offens": 11, "harm": 11, "appli": 11, "both": [6, 11], "within": 11, "space": 11, "individu": 11, "repres": 11, "its": [11, 12], "offici": [11, 12], "mail": 11, "post": [11, 12], "via": 11, "social": 11, "media": 11, "account": 11, "act": 11, "appoint": 11, "onlin": 11, "offlin": 11, "event": 11, "represent": 11, "mai": [2, 11], "further": 11, "defin": 11, "abus": 11, "otherwis": [2, 6, 11], "report": 11, "contact": 11, "team": 11, "review": 11, "complaint": 11, "respond": 11, "wai": 11, "circumst": 11, "oblig": 11, "confidenti": 11, "regard": 11, "incid": 11, "detail": [6, 11, 12], "specif": [2, 11], "polici": 11, "separ": 11, "who": 11, "do": [0, 11], "follow": [2, 6, 11], "good": 11, "faith": 11, "face": 11, "temporari": 11, "repercuss": 11, "determin": 11, "": [0, 11, 12], "leadership": 11, "adapt": 11, "coven": 11, "homepag": 11, "4": 11, "greatli": 12, "appreci": 12, "everi": 12, "littl": 12, "bit": 12, "help": 12, "credit": 12, "alwai": 12, "If": [2, 6, 12], "pleas": [6, 12], "your": [6, 12], "oper": 12, "system": 12, "about": 12, "local": 12, "setup": [12, 14], "might": 12, "troubleshoot": 12, "step": [2, 12], "reproduc": 12, "look": 12, "through": [2, 12], "github": 12, "anyth": 12, "tag": 12, "want": 12, "whoever": 12, "enhanc": 12, "can": [6, 12], "never": 12, "enough": 12, "feel": 12, "part": 12, "doc": [0, 12, 14], "docstr": 12, "even": 12, "web": 12, "blog": 12, "articl": 12, "propos": 12, "explain": 12, "how": 12, "work": [6, 12], "scope": 12, "narrow": 12, "possibl": 12, "easier": 12, "rememb": 12, "volunt": 12, "driven": 12, "readi": 12, "here": 12, "up": 12, "develop": 12, "copi": 12, "instal": 12, "poetri": 12, "git": 12, "similar": 12, "branch": 12, "chang": 12, "checkout": 12, "b": [6, 12], "bugfix": 12, "re": [6, 12], "done": [2, 12], "conform": 12, "pass": 12, "befor": [6, 12], "meet": 12, "addit": 12, "function": 12, "python": 12, "By": 12, "agre": 12, "abid": 12, "term": 12, "To": [6, 13], "import": [6, 13], "print": 13, "attributeerror": [], "traceback": [], "most": [], "recent": [], "call": [], "last": [], "cell": [], "line": [], "3": [], "modul": [], "attribut": [], "md": 14, "py": 14, "licens": 14, "src": 14, "__init__": 14, "datafil": 14, "tutori": [6, 14], "bool": [2, 6], "whether": [2, 6], "origin": 6, "output": 6, "new": 6, "one": 6, "doesn": 6, "t": 6, "input": 6, "unless": 6, "specifi": [2, 6], "invalid": [2, 6], "recommend": 6, "guarante": 6, "ad": 6, "plugin": 0, "provid": [0, 2, 6], "lf": 0, "ff": 0, "well": 0, "fixtur": 0, "control": 0, "see": 0, "between": [2, 6], "speci": 6, "focu": 6, "correspond": 6, "ortholog": 6, "convers": 6, "constant": 6, "design": 6, "genom": 6, "research": 6, "It": [2, 6], "ensur": 6, "accur": 6, "per": 6, "ensdarg00000012345": 6, "ensdarg00000067890": 6, "converted_gen": 6, "For": 6, "usag": 6, "refer": 6, "retain": 6, "miss": 6, "target": 6, "retriev": [2, 6], "read": [2, 6], "subdirectori": 6, "filenotfounderror": 6, "folder": 6, "merg": 6, "zfin_to_ncbi": 6, "zfin_to_ensembl": 6, "unnecessari": 6, "sure": 6, "need": 6, "place": 6, "them": [2, 6], "zfish_human_orthology_v": 6, "process": [2, 6], "zfish_human_orthologi": 6, "duplic": 6, "entri": 6, "string": [2, 6], "preserv": [], "nan": [], "squeez": 6, "valid": [2, 6], "singl": 6, "multipl": 6, "case": 6, "spell": 6, "sensit": 6, "chosen": 6, "human_disease_genes_path": 2, "associ": 2, "obtain": 2, "internet": 2, "access": 2, "code": 2, "against": 2, "tupl": 2, "modifi": 2, "valueerror": 2, "intern": 2, "extract": 2, "mismatch": 2, "relev": 2, "present": 2, "error": 2, "messag": 2, "enter": 2, "prefix": 2, "omit": 2, "intend": 2, "creation": 2, "dure": 2, "time": 2, "exercis": 2, "caution": 2, "store": 2, "carefulli": 2, "involv": 2, "path": 2, "descript": 2, "_download_human_disease_gen": 2, "disease_ids_path": 2, "iter": 2, "human_disease_gen": 2, "_get_genes_for_diseas": 2, "rebuild": 6, "packag": 6, "full": 6, "_build_gene_map": 6, "_build_ortho_map": 6}, "objects": {"": [[5, 0, 0, "-", "danRerLib"]], "danRerLib": [[1, 0, 0, "-", "GO"], [2, 0, 0, "-", "KEGG"], [5, 1, 1, "", "__version__"], [3, 0, 0, "-", "database"], [4, 0, 0, "-", "enrichment"], [6, 0, 0, "-", "mapping"], [7, 0, 0, "-", "settings"], [8, 0, 0, "-", "utils"]], "danRerLib.GO": [[1, 1, 1, "", "GO_BASIC_URL"], [1, 1, 1, "", "GO_IDS_PATH"], [1, 1, 1, "", "GO_NCBI_URL"], [1, 1, 1, "", "GO_PATH_dre"], [1, 1, 1, "", "GO_PATH_dreM"], [1, 1, 1, "", "GO_PATH_hsa"], [1, 1, 1, "", "GO_ZFIN_URL"], [1, 2, 1, "", "_add_to_GO_IDs"], [1, 2, 1, "", "_build_GO_IDs"], [1, 2, 1, "", "_build_dre_mapped_GO"], [1, 2, 1, "", "_build_human_GO"], [1, 2, 1, "", "_build_zebrafish_GO"], [1, 2, 1, "", "_get_ontology"], [1, 2, 1, "", "_process_GOBasic_content"], [1, 2, 1, "", "_write_GOBasic_to_file"], [1, 2, 1, "", "build_all"], [1, 2, 1, "", "check_id_format"], [1, 2, 1, "", "get_genes_in_GO_concept"], [1, 2, 1, "", "id_exists_given_organism"], [1, 2, 1, "", "is_numeric"]], "danRerLib.KEGG": [[2, 2, 1, "", "_build_dre_mapped"], [2, 2, 1, "", "_check_for_organism"], [2, 2, 1, "", "_check_if_pathway_id_exists"], [2, 2, 1, "", "_create_mapped_zebrafish_pathway_ids"], [2, 2, 1, "", "_download_disease_ids"], [2, 2, 1, "", "_download_human_disease_genes"], [2, 2, 1, "", "_download_human_pathway_genes"], [2, 2, 1, "", "_download_human_pathway_ids"], [2, 2, 1, "", "_download_zebrafish_pathway_genes"], [2, 2, 1, "", "_download_zebrafish_pathway_ids"], [2, 2, 1, "", "_get_genes_for_disease"], [2, 2, 1, "", "build_kegg_database"], [2, 1, 1, "", "dre_mapped_dir"], [2, 2, 1, "", "get_genes_in_disease"], [2, 2, 1, "", "get_genes_in_pathway"], [2, 1, 1, "", "human_disease_genes_path"], [2, 1, 1, "", "human_disease_path"], [2, 1, 1, "", "human_pathways_dir"], [2, 1, 1, "", "human_pathways_path"], [2, 1, 1, "", "mapped_zebrafish_pathways_path"], [2, 2, 1, "", "testing"], [2, 1, 1, "", "zebrafish_pathways_dir"], [2, 1, 1, "", "zebrafish_pathways_path"]], "danRerLib.enrichment": [[4, 2, 1, "", "_check_GO_id_format"], [4, 2, 1, "", "_check_concept_ids_GO"], [4, 2, 1, "", "_check_concept_ids_KEGG"], [4, 2, 1, "", "_check_gene_universe"], [4, 2, 1, "", "_check_valid_org"], [4, 2, 1, "", "_id_exists_given_organism"], [4, 2, 1, "", "_is_numeric"], [4, 2, 1, "", "enrich_GO"], [4, 2, 1, "", "enrich_KEGG"], [4, 2, 1, "", "fishers"], [4, 2, 1, "", "investigation"], [4, 2, 1, "", "logistic"], [4, 2, 1, "", "test_GO_enrich"], [4, 2, 1, "", "test_KEGG_enrich"], [4, 2, 1, "", "testing"]], "danRerLib.mapping": [[6, 3, 1, "", "DatabaseNotFoundError"], [6, 3, 1, "", "InvalidGeneTypeError"], [6, 2, 1, "", "_build_gene_mapping"], [6, 2, 1, "", "_build_ortho_mapping"], [6, 2, 1, "", "_check_column_name_matches_id_choice"], [6, 2, 1, "", "_check_valid_gene_id_type_for_orthology"], [6, 2, 1, "", "_check_valid_zebrafish_gene_id_type"], [6, 2, 1, "", "_make_sure_is_pandas_series"], [6, 2, 1, "", "add_mapped_column"], [6, 2, 1, "", "add_mapped_ortholog_column"], [6, 2, 1, "", "convert_ids"], [6, 2, 1, "", "convert_to_human"], [6, 2, 1, "", "convert_to_zebrafish"], [6, 2, 1, "", "get_ortho_ids"]], "danRerLib.settings": [[7, 1, 1, "", "DATABASE_DIR"], [7, 1, 1, "", "ENS_ID"], [7, 1, 1, "", "FILE_DIR"], [7, 1, 1, "", "GO_DATA_DIR"], [7, 1, 1, "", "HUMAN_ID"], [7, 1, 1, "", "KEGG_DATA_DIR"], [7, 1, 1, "", "MASTER_MAPPING_FILE_PATH"], [7, 1, 1, "", "MASTER_ORTHO_FILE_PATH"], [7, 1, 1, "", "NCBI_ID"], [7, 1, 1, "", "RAW_DATA_DIR"], [7, 1, 1, "", "SYMBOL"], [7, 1, 1, "", "VERSION_NUM"], [7, 1, 1, "", "ZFIN_ID"]], "danRerLib.utils": [[8, 2, 1, "", "download_url"], [8, 2, 1, "", "download_url_to_file"], [8, 2, 1, "", "pretty_print_series"], [8, 2, 1, "", "save_data"]]}, "objtypes": {"0": "py:module", "1": "py:data", "2": "py:function", "3": "py:exception"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "data", "Python data"], "2": ["py", "function", "Python function"], "3": ["py", "exception", "Python exception"]}, "titleterms": {"danrerlib": [1, 2, 3, 4, 5, 6, 7, 8], "go": 1, "modul": [1, 2, 4, 6, 7, 8], "content": [1, 2, 4, 5, 6, 7, 8], "function": [1, 2, 4, 6, 8], "attribut": [1, 2, 11], "kegg": 2, "databas": 3, "enrich": 4, "subpackag": 5, "submodul": 5, "packag": 5, "map": 6, "set": 7, "util": 8, "api": 9, "refer": 9, "changelog": 10, "v0": 10, "1": 10, "0": 10, "21": 10, "09": 10, "2023": 10, "code": [11, 12], "conduct": [11, 12], "our": 11, "pledg": 11, "standard": 11, "respons": 11, "scope": 11, "enforc": 11, "contribut": 12, "type": 12, "report": 12, "bug": 12, "fix": 12, "implement": 12, "featur": 12, "write": 12, "document": 12, "submit": 12, "feedback": 12, "get": 12, "start": 12, "pull": 12, "request": 12, "guidelin": 12, "exampl": 13, "usag": 13, "pytest": 0, "cach": 0, "directori": 0}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"Changelog": [[10, "changelog"]], "v0.1.0 (21/09/2023)": [[10, "v0-1-0-21-09-2023"]], "Code of Conduct": [[11, "code-of-conduct"], [12, "code-of-conduct"]], "Our Pledge": [[11, "our-pledge"]], "Our Standards": [[11, "our-standards"]], "Our Responsibilities": [[11, "our-responsibilities"]], "Scope": [[11, "scope"]], "Enforcement": [[11, "enforcement"]], "Attribution": [[11, "attribution"]], "Contributing": [[12, "contributing"]], "Types of Contributions": [[12, "types-of-contributions"]], "Report Bugs": [[12, "report-bugs"]], "Fix Bugs": [[12, "fix-bugs"]], "Implement Features": [[12, "implement-features"]], "Write Documentation": [[12, "write-documentation"]], "Submit Feedback": [[12, "submit-feedback"]], "Get Started!": [[12, "get-started"]], "Pull Request Guidelines": [[12, "pull-request-guidelines"]], "pytest cache directory": [[0, "pytest-cache-directory"]], "Example usage": [[13, "example-usage"]], "danRerLib.GO": [[1, "module-danRerLib.GO"]], "Module Contents": [[1, "module-contents"], [2, "module-contents"], [4, "module-contents"], [6, "module-contents"], [7, "module-contents"], [8, "module-contents"]], "Functions": [[1, "functions"], [2, "functions"], [4, "functions"], [6, "functions"], [8, "functions"]], "Attributes": [[1, "attributes"], [2, "attributes"]], "danRerLib.KEGG": [[2, "module-danRerLib.KEGG"]], "danRerLib.database": [[3, "module-danRerLib.database"]], "danRerLib.enrichment": [[4, "module-danRerLib.enrichment"]], "danRerLib": [[5, "module-danRerLib"]], "Subpackages": [[5, "subpackages"]], "Submodules": [[5, "submodules"]], "Package Contents": [[5, "package-contents"]], "danRerLib.mapping": [[6, "module-danRerLib.mapping"]], "danRerLib.settings": [[7, "module-danRerLib.settings"]], "danRerLib.utils": [[8, "module-danRerLib.utils"]], "API Reference": [[9, "api-reference"]]}, "indexentries": {"go_basic_url (in module danrerlib.go)": [[1, "danRerLib.GO.GO_BASIC_URL"]], "go_ids_path (in module danrerlib.go)": [[1, "danRerLib.GO.GO_IDS_PATH"]], "go_ncbi_url (in module danrerlib.go)": [[1, "danRerLib.GO.GO_NCBI_URL"]], "go_path_dre (in module danrerlib.go)": [[1, "danRerLib.GO.GO_PATH_dre"]], "go_path_drem (in module danrerlib.go)": [[1, "danRerLib.GO.GO_PATH_dreM"]], "go_path_hsa (in module danrerlib.go)": [[1, "danRerLib.GO.GO_PATH_hsa"]], "go_zfin_url (in module danrerlib.go)": [[1, "danRerLib.GO.GO_ZFIN_URL"]], "_add_to_go_ids() (in module danrerlib.go)": [[1, "danRerLib.GO._add_to_GO_IDs"]], "_build_go_ids() (in module danrerlib.go)": [[1, "danRerLib.GO._build_GO_IDs"]], "_build_dre_mapped_go() (in module danrerlib.go)": [[1, "danRerLib.GO._build_dre_mapped_GO"]], "_build_human_go() (in module danrerlib.go)": [[1, "danRerLib.GO._build_human_GO"]], "_build_zebrafish_go() (in module danrerlib.go)": [[1, "danRerLib.GO._build_zebrafish_GO"]], "_get_ontology() (in module danrerlib.go)": [[1, "danRerLib.GO._get_ontology"]], "_process_gobasic_content() (in module danrerlib.go)": [[1, "danRerLib.GO._process_GOBasic_content"]], "_write_gobasic_to_file() (in module danrerlib.go)": [[1, "danRerLib.GO._write_GOBasic_to_file"]], "build_all() (in module danrerlib.go)": [[1, "danRerLib.GO.build_all"]], "check_id_format() (in module danrerlib.go)": [[1, "danRerLib.GO.check_id_format"]], "danrerlib.go": [[1, "module-danRerLib.GO"]], "get_genes_in_go_concept() (in module danrerlib.go)": [[1, "danRerLib.GO.get_genes_in_GO_concept"]], "id_exists_given_organism() (in module danrerlib.go)": [[1, "danRerLib.GO.id_exists_given_organism"]], "is_numeric() (in module danrerlib.go)": [[1, "danRerLib.GO.is_numeric"]], "module": [[1, "module-danRerLib.GO"], [2, "module-danRerLib.KEGG"], [3, "module-danRerLib.database"], [4, "module-danRerLib.enrichment"], [5, "module-danRerLib"], [6, "module-danRerLib.mapping"], [7, "module-danRerLib.settings"], [8, "module-danRerLib.utils"]], "_build_dre_mapped() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._build_dre_mapped"]], "_check_for_organism() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._check_for_organism"]], "_check_if_pathway_id_exists() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._check_if_pathway_id_exists"]], "_create_mapped_zebrafish_pathway_ids() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._create_mapped_zebrafish_pathway_ids"]], "_download_disease_ids() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._download_disease_ids"]], "_download_human_disease_genes() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._download_human_disease_genes"]], "_download_human_pathway_genes() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._download_human_pathway_genes"]], "_download_human_pathway_ids() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._download_human_pathway_ids"]], "_download_zebrafish_pathway_genes() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._download_zebrafish_pathway_genes"]], "_download_zebrafish_pathway_ids() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._download_zebrafish_pathway_ids"]], "_get_genes_for_disease() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG._get_genes_for_disease"]], "build_kegg_database() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.build_kegg_database"]], "danrerlib.kegg": [[2, "module-danRerLib.KEGG"]], "dre_mapped_dir (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.dre_mapped_dir"]], "get_genes_in_disease() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.get_genes_in_disease"]], "get_genes_in_pathway() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.get_genes_in_pathway"]], "human_disease_genes_path (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.human_disease_genes_path"]], "human_disease_path (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.human_disease_path"]], "human_pathways_dir (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.human_pathways_dir"]], "human_pathways_path (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.human_pathways_path"]], "mapped_zebrafish_pathways_path (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.mapped_zebrafish_pathways_path"]], "testing() (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.testing"]], "zebrafish_pathways_dir (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.zebrafish_pathways_dir"]], "zebrafish_pathways_path (in module danrerlib.kegg)": [[2, "danRerLib.KEGG.zebrafish_pathways_path"]], "danrerlib.database": [[3, "module-danRerLib.database"]], "_check_go_id_format() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment._check_GO_id_format"]], "_check_concept_ids_go() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment._check_concept_ids_GO"]], "_check_concept_ids_kegg() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment._check_concept_ids_KEGG"]], "_check_gene_universe() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment._check_gene_universe"]], "_check_valid_org() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment._check_valid_org"]], "_id_exists_given_organism() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment._id_exists_given_organism"]], "_is_numeric() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment._is_numeric"]], "danrerlib.enrichment": [[4, "module-danRerLib.enrichment"]], "enrich_go() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.enrich_GO"]], "enrich_kegg() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.enrich_KEGG"]], "fishers() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.fishers"]], "investigation() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.investigation"]], "logistic() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.logistic"]], "test_go_enrich() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.test_GO_enrich"]], "test_kegg_enrich() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.test_KEGG_enrich"]], "testing() (in module danrerlib.enrichment)": [[4, "danRerLib.enrichment.testing"]], "__version__ (in module danrerlib)": [[5, "danRerLib.__version__"]], "danrerlib": [[5, "module-danRerLib"]], "databasenotfounderror": [[6, "danRerLib.mapping.DatabaseNotFoundError"]], "invalidgenetypeerror": [[6, "danRerLib.mapping.InvalidGeneTypeError"]], "_build_gene_mapping() (in module danrerlib.mapping)": [[6, "danRerLib.mapping._build_gene_mapping"]], "_build_ortho_mapping() (in module danrerlib.mapping)": [[6, "danRerLib.mapping._build_ortho_mapping"]], "_check_column_name_matches_id_choice() (in module danrerlib.mapping)": [[6, "danRerLib.mapping._check_column_name_matches_id_choice"]], "_check_valid_gene_id_type_for_orthology() (in module danrerlib.mapping)": [[6, "danRerLib.mapping._check_valid_gene_id_type_for_orthology"]], "_check_valid_zebrafish_gene_id_type() (in module danrerlib.mapping)": [[6, "danRerLib.mapping._check_valid_zebrafish_gene_id_type"]], "_make_sure_is_pandas_series() (in module danrerlib.mapping)": [[6, "danRerLib.mapping._make_sure_is_pandas_series"]], "add_mapped_column() (in module danrerlib.mapping)": [[6, "danRerLib.mapping.add_mapped_column"]], "add_mapped_ortholog_column() (in module danrerlib.mapping)": [[6, "danRerLib.mapping.add_mapped_ortholog_column"]], "convert_ids() (in module danrerlib.mapping)": [[6, "danRerLib.mapping.convert_ids"]], "convert_to_human() (in module danrerlib.mapping)": [[6, "danRerLib.mapping.convert_to_human"]], "convert_to_zebrafish() (in module danrerlib.mapping)": [[6, "danRerLib.mapping.convert_to_zebrafish"]], "danrerlib.mapping": [[6, "module-danRerLib.mapping"]], "get_ortho_ids() (in module danrerlib.mapping)": [[6, "danRerLib.mapping.get_ortho_ids"]], "database_dir (in module danrerlib.settings)": [[7, "danRerLib.settings.DATABASE_DIR"]], "ens_id (in module danrerlib.settings)": [[7, "danRerLib.settings.ENS_ID"]], "file_dir (in module danrerlib.settings)": [[7, "danRerLib.settings.FILE_DIR"]], "go_data_dir (in module danrerlib.settings)": [[7, "danRerLib.settings.GO_DATA_DIR"]], "human_id (in module danrerlib.settings)": [[7, "danRerLib.settings.HUMAN_ID"]], "kegg_data_dir (in module danrerlib.settings)": [[7, "danRerLib.settings.KEGG_DATA_DIR"]], "master_mapping_file_path (in module danrerlib.settings)": [[7, "danRerLib.settings.MASTER_MAPPING_FILE_PATH"]], "master_ortho_file_path (in module danrerlib.settings)": [[7, "danRerLib.settings.MASTER_ORTHO_FILE_PATH"]], "ncbi_id (in module danrerlib.settings)": [[7, "danRerLib.settings.NCBI_ID"]], "raw_data_dir (in module danrerlib.settings)": [[7, "danRerLib.settings.RAW_DATA_DIR"]], "symbol (in module danrerlib.settings)": [[7, "danRerLib.settings.SYMBOL"]], "version_num (in module danrerlib.settings)": [[7, "danRerLib.settings.VERSION_NUM"]], "zfin_id (in module danrerlib.settings)": [[7, "danRerLib.settings.ZFIN_ID"]], "danrerlib.settings": [[7, "module-danRerLib.settings"]], "danrerlib.utils": [[8, "module-danRerLib.utils"]], "download_url() (in module danrerlib.utils)": [[8, "danRerLib.utils.download_url"]], "download_url_to_file() (in module danrerlib.utils)": [[8, "danRerLib.utils.download_url_to_file"]], "pretty_print_series() (in module danrerlib.utils)": [[8, "danRerLib.utils.pretty_print_series"]], "save_data() (in module danrerlib.utils)": [[8, "danRerLib.utils.save_data"]]}})