Search.setIndex({"docnames": ["_templates/autoapi/index", "autoapi/danrerlib/GO/index", "autoapi/danrerlib/KEGG/index", "autoapi/danrerlib/database/index", "autoapi/danrerlib/enrichment/index", "autoapi/danrerlib/enrichplots/index", "autoapi/danrerlib/index", "autoapi/danrerlib/mapping/index", "autoapi/danrerlib/settings/index", "autoapi/danrerlib/utils/index", "autoapi/index", "devguide/changelog", "devguide/conduct", "devguide/contributing", "devguide/license", "index", "tutorials/functional_enrichment", "tutorials/index", "tutorials/perform_geneid_mapping", "tutorials/perform_orthology_mapping", "userguide/installation", "userguide/introduction", "userguide/usage"], "filenames": ["_templates/autoapi/index.rst", "autoapi/danrerlib/GO/index.rst", "autoapi/danrerlib/KEGG/index.rst", "autoapi/danrerlib/database/index.rst", "autoapi/danrerlib/enrichment/index.rst", "autoapi/danrerlib/enrichplots/index.rst", "autoapi/danrerlib/index.rst", "autoapi/danrerlib/mapping/index.rst", "autoapi/danrerlib/settings/index.rst", "autoapi/danrerlib/utils/index.rst", "autoapi/index.rst", "devguide/changelog.md", "devguide/conduct.md", "devguide/contributing.md", "devguide/license.md", "index.md", "tutorials/functional_enrichment.ipynb", "tutorials/index.md", "tutorials/perform_geneid_mapping.ipynb", "tutorials/perform_orthology_mapping.ipynb", "userguide/installation.md", "userguide/introduction.md", "userguide/usage.md"], "titles": ["API Reference", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.GO</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.KEGG</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.database</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.enrichment</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.enrichplots</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.mapping</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.settings</span></code>", "<code class=\"xref py py-mod docutils literal notranslate\"><span class=\"pre\">danrerlib.utils</span></code>", "API Reference", "Changelog", "Code of Conduct", "Contributing", "License", "Welcome to danRerLib\u2019s Documentation", "Functional Enrichment Analysis", "Tutorials", "Gene ID Mapping", "Orthology Mapping", "Installation", "Introduction", "Usage Principles"], "terms": {"thi": [0, 1, 2, 4, 7, 9, 10, 12, 13, 14, 16, 18, 19, 20, 21, 22], "page": [0, 10, 15, 22], "contain": [0, 1, 2, 4, 7, 10, 16, 18, 19], "document": [0, 1, 2, 4, 7, 10, 14, 18, 21], "The": [1, 2, 4, 7, 9, 12, 13, 14, 16, 18, 19, 20, 22], "gene": [1, 2, 4, 7, 8, 9, 15, 17, 21], "ontologi": [1, 4, 21, 22], "provid": [1, 2, 4, 7, 9, 14, 16, 21], "retriev": [1, 2, 7, 16, 18, 21, 22], "inform": [1, 2, 4, 7, 12, 15, 16, 18, 19, 20, 21, 22], "associ": [1, 2, 14, 16, 22], "biolog": [1, 4, 16, 21, 22], "process": [1, 2, 4, 7, 16, 21, 22], "bp": [1, 4, 16, 21, 22], "molecularfunctionsn": 1, "mf": [1, 4, 16, 21, 22], "cellular": [1, 4, 16, 21, 22], "compon": [1, 4, 16, 21, 22], "cc": [1, 4, 16, 21, 22], "variou": [1, 2, 4, 19], "organ": [1, 2, 4, 7, 9, 16, 18, 19], "includ": [1, 2, 4, 7, 9, 12, 13, 14, 16, 18, 19], "human": [1, 2, 4, 7, 8, 16, 21], "zebrafish": [1, 2, 4, 7, 9, 15, 16, 18, 21], "map": [1, 2, 4, 6, 10, 15, 16, 17, 21, 22], "get_genes_in_go_concept": [1, 16], "specif": [1, 2, 12, 15, 16, 18, 20, 21, 22], "concept": [1, 4, 16, 21], "id": [1, 2, 4, 7, 8, 9, 15, 16, 17, 19, 21], "constant": [1, 2, 4, 7], "ncbi_id": [1, 2, 4, 7, 8, 16, 22], "identifi": [1, 2, 4, 7, 16, 21, 22], "ncbi": [1, 2, 4, 7, 8, 9, 16, 18, 19, 21, 22], "zfin_id": [1, 2, 4, 7, 8, 22], "zfin": [1, 2, 4, 7, 8, 9, 16, 18, 19, 21, 22], "ens_id": [1, 2, 4, 7, 8, 22], "ensembl": [1, 2, 4, 7, 8, 9, 16, 18, 19, 21, 22], "symbol": [1, 2, 4, 7, 8, 9, 16, 18, 19, 21, 22], "human_id": [1, 2, 4, 7, 8, 19], "databas": [1, 2, 4, 7, 16, 18, 19, 21], "rebuild": [1, 2, 7, 21], "build_gene_ontology_databas": 1, "build": [1, 2, 7, 18, 19, 21], "updat": [1, 2, 13, 15, 21], "note": [1, 2, 4, 7, 9, 13, 16, 18, 19], "i": [1, 2, 4, 7, 9, 12, 13, 14, 15, 16, 18, 19, 20, 21, 22], "design": [1, 2, 4, 7, 15, 16, 21], "access": [1, 2], "from": [1, 2, 4, 7, 12, 13, 14, 16, 21, 22], "It": [1, 2, 4, 7, 16, 18, 19, 20, 21, 22], "data": [1, 2, 5, 7, 9, 16, 18, 19, 21], "molecular": [1, 4, 16, 21, 22], "obtain": [1, 2, 14, 16], "either": [1, 2, 16, 18, 19, 20], "through": [1, 2, 13, 16, 18, 19], "read": [1, 2, 7, 18, 19], "pre": [1, 2], "file": [1, 2, 7, 14], "option": [1, 2, 4, 7, 9, 18, 19], "dre": [1, 2, 4, 9, 16], "danio": [1, 2, 15, 16], "rerio": [1, 2, 15, 16], "hsa": [1, 2, 4, 9, 16], "drem": [1, 2, 4, 9, 16], "exampl": [1, 2, 4, 7, 12, 16, 18, 19, 22], "To": [1, 2, 4, 7, 16, 18, 19, 22], "go_gen": 1, "0007582": 1, "For": [1, 2, 4, 7, 15, 16, 18, 19], "detail": [1, 2, 4, 7, 12, 13, 15, 21], "each": [1, 2, 4, 7, 16, 18, 22], "usag": [1, 2, 4, 7, 15, 21], "pleas": [1, 2, 4, 7, 13, 16, 18, 20, 21, 22], "refer": [1, 2, 7, 18, 20, 21, 22], "more": [1, 2, 7, 9, 15, 16, 18, 19, 21, 22], "full": [1, 2, 4, 7, 15, 16, 18], "tutori": [1, 2, 4, 7, 16, 18, 19, 21, 22], "concept_id": [1, 4, 16], "str": [1, 2, 4, 7, 9, 16, 22], "gene_id_typ": [1, 4, 9, 16], "none": [1, 2, 4, 5, 7, 9, 16], "out_data_typ": 1, "type": [1, 4, 7, 9, 16, 18, 19, 21], "pd": [1, 2, 4, 7, 9, 16, 18, 19], "datafram": [1, 2, 4, 7, 9, 16, 18, 19], "do_check": [1, 2], "true": [1, 2, 4, 7, 16, 18, 19], "panda": [1, 2, 7, 16, 18, 19, 20], "seri": [1, 7, 9], "specifi": [1, 2, 4, 7, 9, 18, 19], "paramet": [1, 2, 4, 7, 9, 18], "which": [1, 2, 12, 16, 18, 19], "ar": [1, 2, 4, 7, 9, 12, 13, 16, 18, 19, 20, 22], "should": [1, 2, 13, 16, 18, 19, 20], "same": [1, 2, 16, 18, 19], "character": [1, 2, 19], "desir": [1, 7, 16], "return": [1, 2, 4, 7, 9, 16, 18], "default": [1, 2, 4, 7, 16, 18, 19], "gene_ids_seri": 1, "A": [1, 2, 4, 7, 9, 14, 15, 16, 18, 19, 22], "rais": [1, 7, 9], "valueerror": 1, "If": [1, 2, 4, 7, 9, 13, 15, 16, 18, 19, 20, 21], "doe": [1, 4, 16, 18, 19], "exist": [1, 16, 18, 19], "given": [1, 4, 7, 13, 16, 18, 19], "invalid": [1, 7, 9], "valid": [1, 2, 9, 16], "0001234": 1, "gene_id": 1, "id_exists_given_organ": 1, "check": [1, 7, 9, 13, 15, 18], "code": [1, 4, 15, 16, 18, 19], "e": [1, 4, 12, 16, 19], "g": [1, 4, 16, 19], "bool": [1, 2, 4, 7, 16], "fals": [1, 2, 4, 7, 16, 18, 19], "otherwis": [1, 7, 12, 14, 18, 19], "whether": [1, 2, 4, 7, 14, 16], "look": [1, 13, 16, 18, 19], "match": [1, 7, 18, 19], "column": [1, 7, 16], "perform": [1, 2, 4, 7, 16, 21], "an": [1, 9, 12, 14, 16, 18, 19, 20], "error": [1, 16], "messag": 1, "intend": [1, 2], "creation": [1, 2], "run": [1, 2, 7, 16, 18, 19, 21], "onli": [1, 2, 4, 16, 18, 19], "dure": [1, 2], "version": [1, 2, 12, 13, 16], "mai": [1, 2, 12, 16, 18], "take": [1, 2, 12, 16, 19], "some": [1, 2, 7, 16, 18, 19, 20, 22], "time": [1, 2, 18], "so": [1, 2, 14], "exercis": [1, 2], "caution": [1, 2], "when": [1, 2, 7, 9, 12, 13, 18, 19], "follow": [1, 2, 12, 14, 16, 18, 19, 20, 21, 22], "step": [1, 2, 13, 15, 16, 18, 19, 20], "download": [1, 2, 13, 17, 20], "all": [1, 4, 12, 13, 14, 16, 18, 19, 20], "current": [1, 7, 13, 16, 18, 19, 20, 22], "3": [1, 16, 18, 19, 20], "5": [1, 16, 18, 19], "6": [1, 16, 18, 19], "add": [1, 7, 13, 18, 19], "done": [1, 2, 13, 16, 18, 19], "carefulli": [1, 2], "involv": [1, 2, 16], "fromgo": 1, "pathwai": [2, 4, 18, 19, 21], "diseas": [2, 4, 22], "get_genes_in_pathwai": [2, 16], "get_genes_in_diseas": [2, 16], "build_kegg_databas": 2, "direct": [2, 4, 16], "api": [2, 16, 18, 21, 22], "call": [2, 16, 18, 19], "pathway_gen": 2, "hsa04010": 2, "pathway_id": 2, "org": [2, 4, 9, 16], "interest": [2, 12, 15, 16, 19, 21], "us": [2, 4, 7, 12, 13, 14, 15, 16, 18, 19, 20, 21, 22], "df": 2, "disease_id": 2, "result": [2, 4, 7, 16, 19], "rest": 2, "set": [2, 4, 7, 9, 12, 13, 17, 18, 19, 21, 22], "fetch": 2, "store": [2, 18], "them": [2, 18, 19], "creat": [2, 7, 12, 13, 16, 20], "directori": [2, 7, 18, 19], "offer": [4, 16], "collect": [4, 22], "analys": [4, 21], "These": [4, 19], "allow": [4, 16], "you": [4, 7, 13, 15, 16, 18, 19, 20, 21, 22], "overrepres": [4, 16, 22], "compar": [4, 16, 22], "background": [4, 17], "univers": [4, 16], "support": [4, 7, 13, 16, 18, 19, 22], "method": [4, 16, 18, 19, 22], "make": [4, 7, 12, 13, 15, 20], "versatil": 4, "tool": [4, 20], "uncov": [4, 16], "insight": [4, 16], "enrich_logist": [4, 16], "analysi": [4, 17, 19, 21, 22], "logist": [4, 22], "regress": [4, 22], "ani": [4, 12, 13, 14, 16, 18, 19, 21, 22], "annot": [4, 16, 21, 22], "kegg": [4, 6, 10, 19, 21, 22], "kyoto": [4, 16, 21], "encyclopedia": [4, 16, 21], "genom": [4, 7, 15, 16, 18, 19, 21], "go": [4, 6, 10, 16, 18, 19, 21, 22], "enrich_fish": [4, 16], "fisher": [4, 22], "": [4, 12, 13, 20, 22], "exact": [4, 22], "test": [4, 13, 15, 18, 19, 21, 22], "privat": [4, 12], "_enrich": 4, "launch": [4, 18, 19], "qualiti": 4, "control": [4, 16], "_logist": 4, "statist": [4, 16, 21, 22], "_fisher": 4, "conduct": [4, 15, 17, 21], "analyz": [4, 16], "context": [4, 16], "user": [4, 16], "can": [4, 7, 9, 13, 16, 18, 19, 20, 21, 22], "choos": [4, 16, 19], "differ": [4, 7, 12, 16, 18, 19, 21], "enhanc": [4, 13, 21], "flexibl": [4, 16], "gene_univers": [4, 16], "comprehens": [4, 15, 21], "consult": 4, "also": [4, 16, 18, 19, 22], "find": [4, 16, 19], "demonstr": 4, "list": [4, 7, 9, 16, 19, 22], "non": [4, 16, 18], "sig_gene_cutoff_pvalu": [4, 16], "0": [4, 16, 18, 19, 20], "05": [4, 16], "log2fc_cutoff_valu": [4, 16], "background_gene_set": [4, 16], "sig_conceptid_cutoff_pvalu": [4, 16], "order_by_p_valu": [4, 16], "min_num_genes_in_concept": [4, 16], "10": [4, 16, 18, 19], "include_al": [4, 16], "cut": 4, "off": 4, "free": [4, 12, 13, 14, 18, 19], "p": [4, 5, 16], "valu": [4, 5, 7, 16, 19], "log2fc": [4, 16, 18, 19], "shown": [4, 16, 19], "abov": [4, 14, 16, 18, 19, 22], "combin": [4, 16], "eg": 4, "recommend": [4, 7, 16, 21], "must": [4, 7, 16, 18], "one": [4, 7, 9, 16, 18, 19], "directional_test": [4, 16], "up": [4, 13, 15, 19, 21], "regul": 4, "down": 4, "deplet": 4, "float": [4, 16], "signific": [4, 16, 21], "cutoff": [4, 16], "inclus": [4, 12, 16], "base": [4, 7, 9, 16, 18, 20, 22], "log2": [4, 16], "fold": [4, 16], "chang": [4, 13, 15, 16, 18, 19, 20], "repres": [4, 9, 12, 16], "order": [4, 7, 16, 18, 19], "int": [4, 16], "minimum": [4, 16], "number": [4, 16, 19], "consid": [4, 12, 16], "without": [4, 12, 14, 16, 18, 19], "filter": [4, 16], "belong": [4, 16], "proport": [4, 16], "odd": [4, 16], "ratio": [4, 16], "thei": [4, 12, 13, 18], "come": [4, 16, 18], "depend": [4, 16, 18, 20, 22], "pvalu": [4, 16, 18, 19], "bar_chart": [5, 16], "titl": [5, 16], "xlab": [5, 16], "ylab": 5, "log10": 5, "label_font_s": 5, "16": [5, 19], "max_num_pathwai": [5, 16], "enrich": [6, 10, 15, 17, 21], "enrichplot": [6, 10, 16], "orthologi": [7, 15, 17, 21, 22], "between": [7, 16, 18, 19, 21, 22], "speci": [7, 16], "focu": 7, "convert_id": [7, 18], "convert": [7, 19, 21], "add_mapped_column": [7, 18], "new": [7, 15, 18, 20, 21], "convert_to_human": [7, 19], "correspond": 7, "ortholog": [7, 16, 21], "convert_to_zebrafish": [7, 19], "add_mapped_ortholog_column": [7, 19], "build_gene_map": 7, "master": [7, 18, 19], "convers": [7, 18, 19, 21, 22], "build_ortho_map": 7, "research": [7, 15, 19, 21], "project": [7, 12, 13, 20, 22], "would": [7, 13, 16, 18, 19, 20, 22], "like": [7, 16, 18, 19, 20], "re": [7, 13, 15, 20, 21], "ensur": [7, 21], "your": [7, 13, 16, 18, 19, 20, 21, 22], "dataset": [7, 16], "accur": 7, "name": [7, 9, 13, 16, 18, 19], "per": 7, "requir": [7, 13, 17, 20], "packag": [7, 16, 18, 19, 20, 21], "gene_list": 7, "ensdarg00000012345": 7, "ensdarg00000067890": 7, "converted_gen": 7, "except": [7, 9, 19], "databasenotfounderror": 7, "found": [7, 22], "invalidgenetypeerror": [7, 9], "union": [7, 9], "numpi": [7, 16, 20], "arrai": [7, 16], "id_from": [7, 18, 19], "id_to": [7, 18, 19], "keep_map": [7, 18, 19], "out_format": 7, "format": [7, 9, 13, 16, 18, 19], "np": [7, 16], "keep": [7, 13], "origin": [7, 16, 18], "output": [7, 16, 18, 19], "mapped_gen": [7, 16], "other": [7, 12, 14, 18, 19, 20], "import": [7, 16, 18, 19, 22], "have": [7, 12, 13, 16, 18, 19, 21], "guarante": 7, "column_name_with_id": [7, 18], "keep_old_id": [7, 18], "drop_na": 7, "ha": [7, 16, 18, 19, 22], "doesn": 7, "t": [7, 16, 18, 19, 21], "old": [7, 18], "drop": [7, 18, 19], "row": [7, 18, 19], "na": 7, "ad": [7, 16, 18, 19], "input": [7, 9, 16], "unless": 7, "zfish_gene_typ": [7, 19], "keep_missing_ortho": [7, 19], "retain": 7, "miss": 7, "target": [7, 16], "zebrafish_id": [7, 19], "get_ortho_id": 7, "filenotfounderror": 7, "both": [7, 12, 16, 18, 19], "raw_data": 7, "folder": [7, 18, 19], "work": [7, 13, 16, 18, 19, 20], "subdirectori": [7, 18, 19], "zfin_to_ncbi_v": 7, "version_num": [7, 8], "txt": [7, 16, 18, 19], "zfin_to_ensembl_v": 7, "merg": [7, 14], "save": [7, 18, 19], "master_gene_mapping_file_v": 7, "sure": 7, "befor": [7, 13, 16, 18, 19], "zfish_human_orthology_v": 7, "zfish_human_orthologi": 7, "remov": [7, 12], "unnecessari": 7, "duplic": [7, 19], "entri": [7, 18, 19], "master_ortho_mapping_file_v": 7, "1": [8, 12, 16, 18, 19, 20], "file_dir": 8, "database_dir": 8, "raw_data_dir": 8, "kegg_data_dir": 8, "go_data_dir": 8, "master_mapping_file_path": 8, "master_ortho_file_path": 8, "invalidorganismerror": 9, "check_valid_organ": 9, "chosen": 9, "string": 9, "normalize_organism_nam": 9, "organism_nam": 9, "normal": 9, "check_valid_zebrafish_gene_id_typ": 9, "singl": 9, "multipl": 9, "case": [9, 16], "spell": [9, 18, 19], "sensit": [9, 18, 19], "check_valid_human_gene_id_typ": 9, "normalize_gene_id_typ": 9, "pretty_print_seri": [9, 18, 19], "save_data": [9, 18, 19], "locat": [9, 16, 18, 19, 22], "download_url": 9, "url_to_download": 9, "download_url_to_fil": 9, "output_file_path": 9, "danrerlib": [10, 11, 13, 16, 18, 19, 20, 22], "first": [11, 16, 18, 19], "releas": [11, 13, 21], "In": [12, 16, 18, 19], "foster": 12, "open": [12, 13], "welcom": [12, 13, 21], "environ": [12, 20], "we": [12, 16, 18, 19, 20, 21], "contributor": 12, "maintain": [12, 18, 19], "particip": 12, "commun": [12, 15, 21], "harass": 12, "experi": 12, "everyon": 12, "regardless": [12, 16], "ag": 12, "bodi": 12, "size": 12, "disabl": 12, "ethnic": 12, "gender": 12, "ident": 12, "express": [12, 14, 16], "level": [12, 16], "nation": 12, "person": [12, 14], "appear": [12, 16], "race": 12, "religion": 12, "sexual": 12, "orient": 12, "behavior": 12, "contribut": [12, 15, 20], "posit": [12, 16], "languag": 12, "Being": 12, "respect": [12, 20], "viewpoint": 12, "gracefulli": 12, "accept": 12, "construct": 12, "critic": 12, "focus": [12, 15], "what": [12, 16, 18, 19, 22], "best": [12, 15, 18, 19, 20], "show": [12, 16], "empathi": 12, "toward": 12, "member": 12, "unaccept": 12, "imageri": 12, "unwelcom": 12, "attent": 12, "advanc": 12, "troll": 12, "insult": 12, "derogatori": 12, "comment": 12, "polit": 12, "attack": 12, "public": 12, "publish": [12, 14], "physic": 12, "electron": 12, "address": 12, "explicit": 12, "permiss": [12, 14], "could": [12, 16, 18, 19], "reason": 12, "inappropri": 12, "profession": 12, "clarifi": 12, "expect": [12, 16, 19, 22], "appropri": [12, 13], "fair": 12, "correct": [12, 16, 22], "action": [12, 14], "instanc": 12, "right": [12, 14, 16], "edit": 12, "reject": 12, "commit": [12, 13], "wiki": 12, "issu": [12, 13, 18, 19, 21], "align": [12, 21], "ban": 12, "temporarili": 12, "perman": 12, "deem": 12, "threaten": 12, "offens": 12, "harm": 12, "appli": [12, 16], "within": [12, 16, 22], "space": 12, "individu": 12, "its": [12, 13], "offici": [12, 13, 21], "mail": 12, "post": [12, 13], "via": 12, "social": 12, "media": 12, "account": 12, "act": 12, "appoint": 12, "onlin": 12, "offlin": 12, "event": [12, 14, 16], "represent": [12, 22], "further": [12, 19, 20, 22], "defin": [12, 16, 18, 19], "abus": 12, "report": [12, 19], "contact": [12, 21], "team": 12, "review": [12, 21], "investig": [12, 16, 19], "complaint": 12, "respond": 12, "wai": [12, 18], "circumst": 12, "oblig": 12, "confidenti": 12, "regard": [12, 16], "incid": 12, "polici": 12, "separ": [12, 18, 19], "who": 12, "do": [12, 14, 15, 18, 19], "good": 12, "faith": 12, "face": 12, "temporari": 12, "repercuss": 12, "determin": [12, 16, 19, 22], "leadership": 12, "adapt": 12, "coven": 12, "homepag": 12, "4": [12, 16, 18, 19], "greatli": [13, 16], "appreci": 13, "everi": [13, 16, 19], "littl": 13, "bit": [13, 16], "help": [13, 16], "credit": 13, "alwai": [13, 16, 18, 19], "oper": 13, "system": [13, 16, 21], "about": [13, 15, 16, 18, 20], "local": 13, "setup": 13, "might": [13, 16, 18, 19], "troubleshoot": 13, "reproduc": 13, "github": 13, "anyth": 13, "tag": 13, "want": [13, 15, 16, 18, 19, 21], "whoever": 13, "never": 13, "enough": 13, "feel": [13, 18, 19], "part": 13, "doc": 13, "docstr": 13, "even": [13, 18], "web": 13, "blog": 13, "articl": 13, "propos": 13, "explain": 13, "how": [13, 15, 16, 18, 19, 21], "scope": 13, "narrow": 13, "possibl": [13, 16, 22], "easier": 13, "rememb": [13, 18], "volunt": 13, "driven": 13, "readi": 13, "here": [13, 16, 18, 19], "develop": [13, 16, 22], "copi": [13, 14, 18, 19], "instal": [13, 15, 16, 18, 19, 21], "poetri": 13, "git": 13, "similar": 13, "branch": 13, "checkout": 13, "b": [13, 16, 18, 19], "bugfix": 13, "conform": 13, "pass": 13, "meet": 13, "addit": [13, 16], "function": [13, 15, 17, 18, 19, 21, 22], "python": [13, 15, 16, 18, 19, 20, 21, 22], "By": [13, 18, 19], "agre": 13, "abid": 13, "term": [13, 16], "copyright": 14, "c": [14, 16], "2023": [14, 16, 18, 19], "sdsu": 14, "comput": 14, "toxicologi": 14, "laboratori": 14, "herebi": 14, "grant": 14, "charg": 14, "softwar": [14, 20], "deal": [14, 18, 19], "restrict": 14, "limit": [14, 18], "modifi": 14, "distribut": [14, 16, 21], "sublicens": 14, "sell": 14, "permit": 14, "whom": 14, "furnish": 14, "subject": 14, "condit": [14, 16], "notic": [14, 16, 19], "shall": 14, "substanti": 14, "portion": 14, "THE": 14, "AS": 14, "warranti": 14, "OF": 14, "kind": 14, "OR": 14, "impli": 14, "BUT": 14, "NOT": 14, "TO": 14, "merchant": 14, "fit": 14, "FOR": 14, "particular": [14, 16], "purpos": [14, 17, 20], "AND": 14, "noninfring": 14, "IN": 14, "NO": 14, "author": [14, 16, 18, 19], "holder": 14, "BE": 14, "liabl": 14, "claim": 14, "damag": 14, "liabil": 14, "contract": 14, "tort": 14, "aris": 14, "out": [14, 15, 16, 18, 19], "connect": 14, "WITH": 14, "short": [15, 22], "librari": [15, 16, 18, 19, 20], "toolkit": [15, 22], "transcriptom": [15, 21], "our": [15, 18, 19, 21], "learn": [15, 21], "featur": [15, 22], "explor": [15, 16, 21], "discov": 15, "most": [15, 16, 18, 21], "instruct": [15, 18, 19], "principl": 15, "core": 15, "practic": [15, 20], "effect": [15, 16], "modul": [15, 18, 19, 22], "home": 15, "hand": 15, "changelog": 15, "stai": 15, "date": [15, 16, 18, 19, 21], "latest": [15, 21], "join": 15, "familiar": 15, "yourself": 15, "licens": 15, "ashlei": [16, 18, 19], "schwartz": [16, 18, 19], "octob": 16, "goe": [16, 18, 19], "over": [16, 18, 19, 22], "varieti": [16, 18, 19, 22], "abbrevi": 16, "descript": [16, 18, 19, 22], "link": [16, 18, 19], "interact": 16, "structur": [16, 22], "activ": [16, 18, 20, 22], "product": 16, "convent": [16, 18], "quick": [16, 18, 19, 21], "numer": 16, "prefix": 16, "dre00120": 16, "hsa00120": 16, "bile": 16, "acid": 16, "biosynthesi": 16, "alphanumer": 16, "h00001": 16, "cell": 16, "acut": 16, "lymphoblast": 16, "leukemia": 16, "0007165": 16, "signal": 16, "transduct": 16, "0005634": 16, "nucleu": 16, "0003824": 16, "catalyt": 16, "assess": 16, "predefin": 16, "categori": 16, "categor": 16, "model": 16, "techniqu": 16, "predict": 16, "handl": 16, "divers": 16, "built": [16, 20], "three": [16, 18, 19], "tabl": [16, 18, 19], "ve": 16, "clear": [16, 21], "minor": 16, "typograph": 16, "taxonomi": 16, "util": [16, 18, 19, 20, 22], "see": [16, 18, 19, 20, 22], "matplotlib": [16, 20], "pyplot": 16, "plt": 16, "kei": [16, 18, 19], "benefit": 16, "advantag": [16, 19], "gener": [16, 18, 19, 20], "said": 16, "As": [16, 18, 19], "remind": [16, 18, 19], "Of": [16, 18], "cours": [16, 18], "comparison": 16, "let": [16, 18, 19], "00400": 16, "phenylalanin": 16, "tyrosin": 16, "tryptophan": 16, "hsa00400": 16, "dre00400": 16, "few": [16, 18, 19, 21], "gather": [16, 19], "kegg_id": 16, "human_gen": 16, "137362": 16, "259307": 16, "2": [16, 18, 19, 20], "2805": 16, "2806": 16, "5053": 16, "6898": 16, "dre_gen": 16, "335974": 16, "337166": 16, "378962": 16, "406330": 16, "406688": 16, "561410": 16, "791730": 16, "sent": 16, "omit": 16, "below": [16, 18, 19, 20], "where": [16, 18, 19, 22], "get": [16, 18], "Not": 16, "25": 16, "4297": 16, "4299": 16, "6929": 16, "5087": 16, "861": 16, "4609": 16, "7": [16, 18, 19], "64109": 16, "8": [16, 18, 19, 20], "5079": 16, "100000720": 16, "557048": 16, "100537394": 16, "664768": 16, "30310": 16, "58138": 16, "570960": 16, "58126": 16, "393141": 16, "9": [16, 18, 19, 20], "30686": 16, "100294510": 16, "11": [16, 18], "60636": 16, "describ": 16, "equal": 16, "becaus": [16, 19], "There": [16, 18, 19], "0033554": 16, "respons": 16, "stress": 16, "zdb": [16, 18, 19], "030131": [16, 18, 19], "8638": 16, "091204": 16, "246": [16, 18], "040718": 16, "255": 16, "050320": [16, 19], "35": 16, "a0a2r8s050": 16, "6701": 16, "040825": 16, "1096": 16, "9531": 16, "030826": 16, "18": 16, "030827": 16, "030326": 16, "power": 16, "entir": [16, 18, 19], "abl": 16, "section": 16, "ll": 16, "scenario": [16, 18, 19], "small": [16, 18, 19], "larger": 16, "scale": 16, "plot": 16, "just": [16, 18, 19, 21], "notebook": 16, "displai": 16, "ipython": 16, "markdown": 16, "differenti": 16, "significantli": 16, "alter": 16, "therefor": [16, 18, 19, 20], "detect": 16, "studi": 16, "adjust": 16, "anoth": [16, 18, 19], "metric": 16, "minim": 16, "neg": 16, "second": 16, "third": 16, "_": 16, "publicli": 16, "avail": [16, 19], "omnibu": 16, "gse": 16, "chemic": 16, "tpp": 16, "whole": 16, "embryo": 16, "rna": 16, "sequenc": 16, "preform": 16, "amount": 16, "ideal": 16, "reveal": 16, "relationship": 16, "conting": 16, "begin": [16, 18, 22], "hline": 16, "text": 16, "d": 16, "end": [16, 19], "neither": 16, "nor": 16, "formula": 16, "frac": 16, "bc": 16, "x": [16, 18, 19], "leq": 16, "binom": 16, "n": [16, 19, 20], "total": 16, "calcul": 16, "observ": 16, "extrem": [16, 18], "sum_": 16, "low": 16, "suggest": 16, "unlik": 16, "occur": 16, "random": 16, "chanc": [16, 22], "indic": 16, "potenti": 16, "discuss": 16, "load": 16, "gene_universe_tpp": 16, "read_csv": [16, 18, 19], "test_data": [16, 18, 19], "tpp_de": 16, "sep": [16, 18, 19], "sneak": 16, "peak": 16, "element": [16, 18, 19], "head": [16, 18, 19], "logfc": [16, 18, 19], "100000006": [16, 18, 19], "792615": [16, 18, 19], "115009": [16, 18, 19], "100000009": [16, 18], "607285": [16, 18], "144714": [16, 18], "100000026": [16, 18], "021338": [16, 18], "603871": [16, 18], "100000030": 16, "007880": 16, "083141": 16, "100000044": [16, 18, 19], "015286": [16, 18, 19], "803879": [16, 18, 19], "necessari": 16, "choic": [16, 18, 19], "wish": [16, 22], "database_choic": 16, "visual": [16, 18, 19], "wa": [16, 18, 20], "mean": [16, 18, 19], "count": 16, "num_alt": 16, "len": [16, 19], "num_enrich": 16, "num_deplet": 16, "print": [16, 18, 19], "f": [16, 18, 19], "28": [16, 18], "captur": 16, "top": 16, "to_markdown": 16, "sig": 16, "dre00030": 16, "pentos": 16, "phosphat": 16, "29": 16, "15": 16, "517241": 16, "87263": 16, "13465e": 16, "dre04512": 16, "ecm": 16, "receptor": 16, "101": 16, "346535": 16, "41698": 16, "11217e": 16, "dre04141": 16, "protein": 16, "endoplasm": 16, "reticulum": 16, "190": 16, "54": 16, "284211": 16, "81138": 16, "000419319": 16, "dre04010": 16, "mapk": 16, "377": 16, "95": 16, "251989": 16, "54043": 16, "000544843": 16, "dre04510": 16, "focal": 16, "adhes": 16, "264": 16, "69": 16, "261364": 16, "61511": 16, "00121163": 16, "bar": 16, "chart": 16, "enriched_result": 16, "20": 16, "interpret": 16, "relev": 16, "expos": 16, "87": 16, "veri": [16, 18], "000041": 16, "strong": 16, "assum": 16, "test_direct": 16, "upregulated_result": 16, "num_upregul": 16, "13": [16, 18, 20], "64": 16, "169761": 16, "01456": 16, "97328e": 16, "06": 16, "dre04144": 16, "endocytosi": 16, "325": 16, "47": 16, "144615": 16, "65448": 16, "00175717": 16, "dre04140": 16, "autophagi": 16, "anim": 16, "180": [16, 19], "155556": 16, "79722": 16, "00509073": 16, "dre04012": 16, "erbb": 16, "104": 16, "173077": 16, "03876": 16, "00756464": 16, "dre04070": 16, "phosphatidylinositol": 16, "117": [16, 19], "19": 16, "162393": 16, "88829": 16, "0121393": 16, "repeat": [16, 18, 19], "downregulated_result": 16, "num_downregul": 16, "40": 16, "threshold": 16, "definit": [16, 22], "condis": 16, "factor": 16, "relat": 16, "common": [16, 18], "significan": 16, "incorpoor": 16, "enrichment_result": 16, "lot": [16, 18, 19], "fewer": 16, "v": 16, "stricter": 16, "those": [16, 19], "longer": [16, 18], "known": [16, 18, 19, 22], "better": [16, 19], "than": [16, 18, 19, 20], "sartor": 16, "et": 16, "al": 16, "2009": 16, "log": 16, "linear": 16, "measur": 16, "breakdown": 16, "mention": [16, 20], "explan": 16, "left": [16, 19], "alpha": [16, 19], "beta": [16, 19], "cdot": 16, "logarithm": 16, "probabl": 16, "intercept": 16, "slope": 16, "explanatori": 16, "variabl": [16, 18, 19], "unit": 16, "increas": 16, "decreas": 16, "exp": 16, "wald": 16, "evid": 16, "zero": 16, "w": 16, "chi": 16, "degre": 16, "freedom": 16, "null": 16, "estim": 16, "maximum": 16, "likelihood": 16, "hat": 16, "iter": 16, "weight": 16, "least": 16, "squar": 16, "iwl": 16, "algorithm": 16, "rank": 16, "aim": 16, "strength": 16, "signifi": 16, "56": 16, "14": 16, "42": 16, "dre03013": 16, "nucleocytoplasm": 16, "transport": 16, "100": 16, "26": [16, 20], "691623": 16, "00369e": 16, "dre01200": 16, "carbon": 16, "metabol": 16, "124": 16, "30": 16, "241935": 16, "708663": 16, "32419e": 16, "dre00970": 16, "aminoacyl": 16, "trna": 16, "51": 16, "12": [16, 18], "235294": 16, "680409": 16, "18624e": 16, "07": 16, "dre04110": 16, "cycl": 16, "177": 16, "163842": 16, "783967": 16, "68528e": 16, "dre00010": 16, "glycolysi": 16, "gluconeogenesi": 16, "75": 16, "186667": 16, "725094": 16, "87168e": 16, "39": 16, "34": [16, 18], "46137": 16, "93196e": 16, "2104": 16, "29626": 16, "32985e": 16, "dre00860": 16, "porphyrin": 16, "25641": 16, "3912": 16, "70928e": 16, "16901": 16, "000106494": 16, "were": 16, "mani": 16, "been": [16, 19, 22], "leverag": 16, "knowledg": 16, "well": [16, 19], "less": 16, "transfer": 16, "humamn": 16, "extens": [16, 18, 19], "experiment": 16, "bridg": 16, "gap": 16, "understand": [16, 18, 19], "stand": 16, "organism_for_orthologi": 16, "81": 16, "17": 16, "significanli": 16, "fun": 16, "_get_total_number_kegg": 16, "355": 16, "while": [16, 18, 19], "drem01200": 16, "697511": 16, "15744e": 16, "drem03013": 16, "94": 16, "22": 16, "234043": 16, "704551": 16, "54716e": 16, "08": 16, "drem00970": 16, "38": 16, "263158": 16, "654383": 16, "35405e": 16, "drem00010": 16, "68": 16, "205882": 16, "705103": 16, "35445e": 16, "drem00051": 16, "fructos": 16, "mannos": 16, "236842": 16, "677058": 16, "35243e": 16, "under": [16, 21, 22], "irregardless": 16, "being": 16, "two": [16, 18, 19], "treatment": 16, "higher": 16, "sai": 16, "opposit": 16, "lower": 16, "major": 16, "group": 16, "exhibit": 16, "conclud": [16, 18, 19], "summari": [16, 18, 19], "m": [16, 19, 20], "leikauf": 16, "medvedov": 16, "lrpath": 16, "approach": 16, "bioinformat": 16, "oxford": 16, "england": 16, "211": 16, "217": 16, "execut": 17, "conclus": 17, "glossari": 17, "septemb": [18, 19], "simpli": 18, "form": 18, "confus": 18, "start": [18, 19, 22], "zebafish": [18, 19], "011219": [18, 19], "integ": [18, 19, 22], "manag": [18, 19, 20, 22], "entrez": [18, 19, 22], "140634": [18, 19], "refseq": [18, 19], "cyp1a": [18, 19], "nomenclatur": [18, 19, 22], "ensdar": [18, 19, 22], "ensdarg00000098315": [18, 19], "sampl": [18, 19], "csv": [18, 19], "tsv": [18, 19], "past": [18, 19], "rel": [18, 19], "path": [18, 19], "example_diff_express_data": [18, 19], "need": [18, 19, 20], "larg": [18, 19], "foundat": [18, 19], "absolut": [18, 19], "simplest": 18, "sinc": 18, "excel": [18, 19], "fil": 18, "especi": [18, 19], "list_of_gene_id": 18, "100000252": 18, "100000750": 18, "100001198": 18, "100001260": 18, "100002225": 18, "100002263": [18, 19], "100002756": 18, "100003223": 18, "100007521": 18, "100149273": 18, "100149794": 18, "100170795": 18, "100321746": 18, "100329897": 18, "100330617": 18, "tell": [18, 19], "program": [18, 19], "don": [18, 19, 21], "worri": 18, "know": [18, 19], "made": 18, "mistak": 18, "current_gene_id_typ": 18, "desired_gene_id_typ": 18, "onc": 18, "converted_id": 18, "shell": [18, 19], "fine": 18, "idea": 18, "print_series_pretti": [18, 19], "1904": 18, "3404": 18, "030325": 18, "030616": 18, "609": 18, "040426": [18, 19], "743": 18, "050309": 18, "071009": 18, "080219": 18, "080723": 18, "44": 18, "081223": 18, "090313": 18, "141": 18, "091117": 18, "110309": 18, "120215": 18, "92": 18, "rather": [18, 19, 20], "previous": [18, 19], "file_nam": [18, 19], "out_data": [18, 19], "easili": [18, 19, 21], "data_file_path": [18, 19], "ncbi_gene_list": 18, "30301": 18, "30459": 18, "57924": 18, "564772": 18, "794572": 18, "000616": 18, "050419": 18, "85": 18, "080415": 18, "991229": 18, "manner": 18, "now": 18, "quit": 18, "sometim": 18, "along": 18, "next": 18, "am": [18, 19], "my": [18, 19, 20], "converst": 18, "earlier": 18, "converted_id_t": 18, "automat": 18, "besid": 18, "initi": 18, "methodologi": 18, "great": 18, "properli": [18, 22], "sort": 18, "y": [18, 19], "z": [18, 19], "03": [18, 19], "essenti": [18, 19, 21], "geneid": [18, 19], "often": [18, 19], "sub": [18, 19], "realli": [18, 19], "stat": [18, 19], "line": [18, 19], "col": [18, 19], "shape": [18, 19], "5464": [18, 19], "give": [18, 19], "new_data": [18, 19], "5654": 18, "121214": 18, "60": 18, "100000085": [18, 19], "1312": 18, "264762": [18, 19], "267360": [18, 19], "converted_datafram": 18, "test_set_invalid_col_nam": 18, "around": 18, "exactli": 18, "seen": 18, "130530": 18, "778": 18, "120823": 18, "successfulli": 18, "though": [18, 19], "final": 18, "anymor": 18, "vice": 19, "versa": 19, "four": 19, "1543": 19, "nci": 19, "hsa04911": 19, "dre04910": 19, "alreadi": [19, 20], "mechan": 19, "yet": 19, "approxim": 19, "70": 19, "list_of_zfish_id": 19, "chose": 19, "suffic": 19, "081113": 19, "031002": 19, "50": 19, "120104": 19, "000607": 19, "060503": 19, "934": 19, "72": 19, "3147": 19, "2551": 19, "aka": 19, "4293": 19, "81553": 19, "26207": 19, "605": 19, "4861": 19, "55022": 19, "221143": 19, "save_seri": 19, "again": 19, "list_of_human_id": 19, "55585": 19, "23191": 19, "4192": 19, "5686": 19, "197021": 19, "390": 19, "344805": 19, "2623": 19, "zfish_desired_gene_id_typ": 19, "mdkb": 19, "cyfip1": 19, "psma5": 19, "lctla": 19, "rnd3b": 19, "ube2q1": 19, "gata1a": 19, "mdka": 19, "lctlb": 19, "gata1b": 19, "tmprss7": 19, "closer": 19, "state": 19, "fact": 19, "file_path": 19, "human_ids_with_map": 19, "nan": 19, "downstream": 19, "isn": 19, "dropna": 19, "Or": 19, "ran": 19, "nearli": 19, "kept": 19, "04910": 19, "had": 19, "recal": 19, "dre05910_human_gen": 19, "04911": 19, "86": 19, "desired_zfish_gene_typ": 19, "zfish_id": 19, "64272": 19, "64269": 19, "64267": 19, "dtype": 19, "object": 19, "112": 19, "someth": 19, "140615": 19, "9415": 19, "add_mapped_ortholg_column": 19, "84561": 19, "968": 19, "ortholog_datafram": 19, "went": 19, "113": 19, "proper": 20, "own": 20, "guidelin": [20, 21], "virtual": 20, "complic": 20, "anaconda": 20, "elsewher": 20, "decis": 20, "On": 20, "window": 20, "maco": 20, "linux": 20, "app": 20, "task": [20, 21], "env": 20, "venv": 20, "mkdir": 20, "src": 20, "cd": 20, "bin": 20, "sourc": 20, "assist": 21, "resourc": 21, "guid": 21, "minut": 21, "eas": 21, "simplifi": 21, "regularli": 21, "improv": 21, "encount": 21, "question": 21, "hesit": 21, "u": 21, "mit": 21, "drl": 22, "typic": 22, "introduct": 22, "referenc": 22, "IF": 22, "websit": 22, "network": 22, "alphabet": 22, "mor": 22, "uniqu": 22, "curat": 22, "biochem": 22, "catalog": 22, "standard": 22, "vocabulari": 22, "subcellular": 22, "underrepres": 22, "info": 22}, "objects": {"": [[6, 0, 0, "-", "danrerlib"]], "danrerlib": [[1, 0, 0, "-", "GO"], [2, 0, 0, "-", "KEGG"], [3, 0, 0, "-", "database"], [4, 0, 0, "-", "enrichment"], [5, 0, 0, "-", "enrichplots"], [7, 0, 0, "-", "mapping"], [8, 0, 0, "-", "settings"], [9, 0, 0, "-", "utils"]], "danrerlib.GO": [[1, 1, 1, "", "build_gene_ontology_database"], [1, 1, 1, "", "get_genes_in_GO_concept"], [1, 1, 1, "", "id_exists_given_organism"]], "danrerlib.KEGG": [[2, 1, 1, "", "build_kegg_database"], [2, 1, 1, "", "get_genes_in_disease"], [2, 1, 1, "", "get_genes_in_pathway"]], "danrerlib.enrichment": [[4, 1, 1, "", "enrich_fishers"], [4, 1, 1, "", "enrich_logistic"]], "danrerlib.enrichplots": [[5, 1, 1, "", "bar_chart"]], "danrerlib.mapping": [[7, 2, 1, "", "DatabaseNotFoundError"], [7, 2, 1, "", "InvalidGeneTypeError"], [7, 1, 1, "", "add_mapped_column"], [7, 1, 1, "", "add_mapped_ortholog_column"], [7, 1, 1, "", "build_gene_mapping"], [7, 1, 1, "", "build_ortho_mapping"], [7, 1, 1, "", "convert_ids"], [7, 1, 1, "", "convert_to_human"], [7, 1, 1, "", "convert_to_zebrafish"], [7, 1, 1, "", "get_ortho_ids"]], "danrerlib.settings": [[8, 3, 1, "", "DATABASE_DIR"], [8, 3, 1, "", "ENS_ID"], [8, 3, 1, "", "FILE_DIR"], [8, 3, 1, "", "GO_DATA_DIR"], [8, 3, 1, "", "HUMAN_ID"], [8, 3, 1, "", "KEGG_DATA_DIR"], [8, 3, 1, "", "MASTER_MAPPING_FILE_PATH"], [8, 3, 1, "", "MASTER_ORTHO_FILE_PATH"], [8, 3, 1, "", "NCBI_ID"], [8, 3, 1, "", "RAW_DATA_DIR"], [8, 3, 1, "", "SYMBOL"], [8, 3, 1, "", "VERSION_NUM"], [8, 3, 1, "", "ZFIN_ID"]], "danrerlib.utils": [[9, 2, 1, "", "InvalidGeneTypeError"], [9, 2, 1, "", "InvalidOrganismError"], [9, 1, 1, "", "check_valid_human_gene_id_type"], [9, 1, 1, "", "check_valid_organism"], [9, 1, 1, "", "check_valid_zebrafish_gene_id_type"], [9, 1, 1, "", "download_url"], [9, 1, 1, "", "download_url_to_file"], [9, 1, 1, "", "normalize_gene_id_type"], [9, 1, 1, "", "normalize_organism_name"], [9, 1, 1, "", "pretty_print_series"], [9, 1, 1, "", "save_data"]]}, "objtypes": {"0": "py:module", "1": "py:function", "2": "py:exception", "3": "py:data"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"], "2": ["py", "exception", "Python exception"], "3": ["py", "data", "Python data"]}, "titleterms": {"api": [0, 10, 15], "refer": [0, 10, 15, 16], "danrerlib": [1, 2, 3, 4, 5, 6, 7, 8, 9, 15, 21], "go": 1, "modul": [1, 2, 4, 5, 7, 8, 9, 16], "content": [1, 2, 4, 5, 7, 8, 9], "function": [1, 2, 4, 5, 7, 9, 16], "kegg": [2, 16], "databas": [3, 22], "enrich": [4, 16, 22], "enrichplot": 5, "submodul": 6, "map": [7, 18, 19], "set": [8, 16], "util": 9, "changelog": 11, "v0": 11, "1": 11, "0": 11, "21": 11, "09": 11, "2023": 11, "code": [12, 13], "conduct": [12, 13, 16], "our": 12, "pledg": 12, "standard": 12, "respons": 12, "scope": 12, "enforc": 12, "attribut": 12, "contribut": [13, 21], "type": [13, 22], "report": 13, "bug": 13, "fix": 13, "implement": 13, "featur": [13, 21], "write": 13, "document": [13, 15], "submit": 13, "feedback": [13, 21], "get": [13, 15, 19, 21], "start": [13, 15, 21], "pull": 13, "request": 13, "guidelin": 13, "licens": [14, 21], "mit": 14, "welcom": 15, "": [15, 16], "introduct": [15, 21], "user": [15, 20], "guid": [15, 20], "tutori": [15, 17], "develop": [15, 20], "analysi": 16, "purpos": [16, 18, 19], "background": [16, 18, 19], "requir": [16, 18, 19], "download": 16, "gene": [16, 18, 19, 22], "pathwai": [16, 22], "diseas": 16, "ontologi": 16, "overview": 16, "fisher": 16, "exact": 16, "test": 16, "overrepresent": 16, "primari": 16, "uniqu": 16, "paramet": 16, "deplet": 16, "upregul": 16, "downregul": 16, "advanc": [16, 20], "option": 16, "logist": 16, "regress": 16, "identif": 16, "up": 16, "down": 16, "regul": 16, "via": 16, "orthologi": [16, 19], "glossari": 16, "conclus": [16, 18, 19], "id": [18, 22], "execut": [18, 19], "simpl": [18, 19], "case": [18, 19], "convert": 18, "list": 18, "from": [18, 19], "file": [18, 19], "keep": [18, 19], "column": [18, 19], "larger": [18, 19], "dataset": [18, 19], "special": 18, "ortholog": 19, "zebrafish": [19, 22], "human": [19, 22], "instal": 20, "quick": 20, "conda": 20, "pypi": 20, "pip": 20, "version": 20, "begin": 20, "recommend": 20, "method": 20, "kei": 21, "why": 21, "support": 21, "usag": 22, "principl": 22, "workflow": 22}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx": 57}, "alltitles": {"API Reference": [[0, "api-reference"], [10, "api-reference"], [15, "api-reference"]], "danrerlib.GO": [[1, "module-danrerlib.GO"]], "Module Contents": [[1, "module-contents"], [2, "module-contents"], [4, "module-contents"], [5, "module-contents"], [7, "module-contents"], [8, "module-contents"], [9, "module-contents"]], "Functions": [[1, "functions"], [2, "functions"], [4, "functions"], [5, "functions"], [7, "functions"], [9, "functions"]], "danrerlib.KEGG": [[2, "module-danrerlib.KEGG"]], "danrerlib.database": [[3, "module-danrerlib.database"]], "danrerlib.enrichment": [[4, "module-danrerlib.enrichment"]], "danrerlib.enrichplots": [[5, "module-danrerlib.enrichplots"]], "danrerlib": [[6, "module-danrerlib"]], "Submodules": [[6, "submodules"]], "danrerlib.mapping": [[7, "module-danrerlib.mapping"]], "danrerlib.settings": [[8, "module-danrerlib.settings"]], "danrerlib.utils": [[9, "module-danrerlib.utils"]], "Changelog": [[11, "changelog"]], "v0.1.0 (21/09/2023)": [[11, "v0-1-0-21-09-2023"]], "Code of Conduct": [[12, "code-of-conduct"], [13, "code-of-conduct"]], "Our Pledge": [[12, "our-pledge"]], "Our Standards": [[12, "our-standards"]], "Our Responsibilities": [[12, "our-responsibilities"]], "Scope": [[12, "scope"]], "Enforcement": [[12, "enforcement"]], "Attribution": [[12, "attribution"]], "Contributing": [[13, "contributing"], [21, "contributing"]], "Types of Contributions": [[13, "types-of-contributions"]], "Report Bugs": [[13, "report-bugs"]], "Fix Bugs": [[13, "fix-bugs"]], "Implement Features": [[13, "implement-features"]], "Write Documentation": [[13, "write-documentation"]], "Submit Feedback": [[13, "submit-feedback"]], "Get Started!": [[13, "get-started"]], "Pull Request Guidelines": [[13, "pull-request-guidelines"]], "License": [[14, "license"], [21, "license"]], "MIT License": [[14, "mit-license"]], "Welcome to danRerLib\u2019s Documentation": [[15, "welcome-to-danrerlib-s-documentation"]], "Introduction": [[15, "introduction"], [21, "introduction"]], "Getting Started": [[15, "getting-started"], [21, "getting-started"]], "User\u2019s Guide": [[15, "user-s-guide"]], "Tutorials": [[15, "tutorials"], [17, "tutorials"]], "Developer Guide": [[15, "developer-guide"]], "Functional Enrichment Analysis": [[16, "functional-enrichment-analysis"]], "Purpose and Background": [[16, "purpose-and-background"], [18, "purpose-and-background"], [19, "purpose-and-background"]], "Requirements": [[16, "requirements"], [18, "requirements"], [19, "requirements"]], "Download Gene Sets": [[16, "download-gene-sets"]], "KEGG Pathway": [[16, "kegg-pathway"]], "KEGG Disease": [[16, "kegg-disease"]], "Gene Ontology": [[16, "gene-ontology"]], "Conduct Functional Enrichment Analysis": [[16, "conduct-functional-enrichment-analysis"]], "Enrichment Module Overview": [[16, "enrichment-module-overview"]], "Functional Enrichment with Fisher\u2019s Exact Test": [[16, "functional-enrichment-with-fisher-s-exact-test"]], "Overview of Fisher\u2019s Exact Test for Overrepresentation": [[16, "overview-of-fisher-s-exact-test-for-overrepresentation"]], "Primary Function and Unique Parameters": [[16, "primary-function-and-unique-parameters"], [16, "id1"]], "Conduct Fishers Exact Test for Enrichment Analysis (enriched/depleted)": [[16, "conduct-fishers-exact-test-for-enrichment-analysis-enriched-depleted"]], "Testing for Upregulated Pathways": [[16, "testing-for-upregulated-pathways"]], "Testing for Downregulated Pathways": [[16, "testing-for-downregulated-pathways"]], "Advanced Options": [[16, "advanced-options"]], "Functional Enrichment with Logistic Regression": [[16, "functional-enrichment-with-logistic-regression"]], "Overview of Logistic Regression for Enrichment Analysis": [[16, "overview-of-logistic-regression-for-enrichment-analysis"]], "Logistic Regression for the Identification of Up and Down Regulated Pathways": [[16, "logistic-regression-for-the-identification-of-up-and-down-regulated-pathways"]], "Conduct Logistic Regression for Enrichment Analysis (enriched/depleted)": [[16, "conduct-logistic-regression-for-enrichment-analysis-enriched-depleted"]], "Functional Enrichment via Orthology": [[16, "functional-enrichment-via-orthology"]], "Conduct Logistic Regression for Enrichment Analysis": [[16, "conduct-logistic-regression-for-enrichment-analysis"]], "Glossary": [[16, "glossary"]], "Conclusion": [[16, "conclusion"], [18, "conclusion"], [19, "conclusion"]], "References": [[16, "references"]], "Gene ID Mapping": [[18, "gene-id-mapping"]], "Execute Mappings": [[18, "execute-mappings"]], "Simple Case: Convert a list of Gene IDs": [[18, "simple-case-convert-a-list-of-gene-ids"]], "Simple Case: Convert a list of Gene IDs From a File": [[18, "simple-case-convert-a-list-of-gene-ids-from-a-file"]], "Simple Case: Convert a list of Gene IDs and Keep Mapping": [[18, "simple-case-convert-a-list-of-gene-ids-and-keep-mapping"]], "Convert Gene IDs in a Column of a Larger Dataset": [[18, "convert-gene-ids-in-a-column-of-a-larger-dataset"]], "Special Cases": [[18, "special-cases"]], "Orthology Mapping": [[19, "orthology-mapping"]], "Execute Orthology Mappings": [[19, "execute-orthology-mappings"]], "Simple Case: Get orthologs": [[19, "simple-case-get-orthologs"]], "Zebrafish to Human": [[19, "zebrafish-to-human"], [19, "id1"], [19, "id3"]], "Human to Zebrafish": [[19, "human-to-zebrafish"], [19, "id2"], [19, "id4"]], "Simple Case: Get Orthologs and Keep Mapping": [[19, "simple-case-get-orthologs-and-keep-mapping"]], "Get Orthologs for Genes from File": [[19, "get-orthologs-for-genes-from-file"]], "Get Orthologs for a Column of a Larger Dataset": [[19, "get-orthologs-for-a-column-of-a-larger-dataset"]], "Installation": [[20, "installation"]], "Quick Install": [[20, "quick-install"]], "Conda": [[20, "conda"]], "PyPi/pip": [[20, "pypi-pip"]], "Development Version": [[20, "development-version"]], "Installation Guide": [[20, "installation-guide"]], "Beginning users": [[20, "beginning-users"]], "Advanced users (recommended method)": [[20, "advanced-users-recommended-method"]], "Key Features": [[21, "key-features"]], "Why danRerLib?": [[21, "why-danrerlib"]], "Support and Feedback": [[21, "support-and-feedback"]], "Usage Principles": [[22, "usage-principles"]], "Workflow": [[22, "workflow"]], "Zebrafish Gene ID Types": [[22, "zebrafish-gene-id-types"]], "Human Gene ID Type": [[22, "human-gene-id-type"]], "Pathway Databases": [[22, "pathway-databases"]], "Enrichment Types": [[22, "enrichment-types"]]}, "indexentries": {"build_gene_ontology_database() (in module danrerlib.go)": [[1, "danrerlib.GO.build_gene_ontology_database"]], "danrerlib.go": [[1, "module-danrerlib.GO"]], "get_genes_in_go_concept() (in module danrerlib.go)": [[1, "danrerlib.GO.get_genes_in_GO_concept"]], "id_exists_given_organism() (in module danrerlib.go)": [[1, "danrerlib.GO.id_exists_given_organism"]], "module": [[1, "module-danrerlib.GO"], [2, "module-danrerlib.KEGG"], [3, "module-danrerlib.database"], [4, "module-danrerlib.enrichment"], [5, "module-danrerlib.enrichplots"], [6, "module-danrerlib"], [7, "module-danrerlib.mapping"], [8, "module-danrerlib.settings"], [9, "module-danrerlib.utils"]], "build_kegg_database() (in module danrerlib.kegg)": [[2, "danrerlib.KEGG.build_kegg_database"]], "danrerlib.kegg": [[2, "module-danrerlib.KEGG"]], "get_genes_in_disease() (in module danrerlib.kegg)": [[2, "danrerlib.KEGG.get_genes_in_disease"]], "get_genes_in_pathway() (in module danrerlib.kegg)": [[2, "danrerlib.KEGG.get_genes_in_pathway"]], "danrerlib.database": [[3, "module-danrerlib.database"]], "danrerlib.enrichment": [[4, "module-danrerlib.enrichment"]], "enrich_fishers() (in module danrerlib.enrichment)": [[4, "danrerlib.enrichment.enrich_fishers"]], "enrich_logistic() (in module danrerlib.enrichment)": [[4, "danrerlib.enrichment.enrich_logistic"]], "bar_chart() (in module danrerlib.enrichplots)": [[5, "danrerlib.enrichplots.bar_chart"]], "danrerlib.enrichplots": [[5, "module-danrerlib.enrichplots"]], "danrerlib": [[6, "module-danrerlib"]], "databasenotfounderror": [[7, "danrerlib.mapping.DatabaseNotFoundError"]], "invalidgenetypeerror": [[7, "danrerlib.mapping.InvalidGeneTypeError"], [9, "danrerlib.utils.InvalidGeneTypeError"]], "add_mapped_column() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.add_mapped_column"]], "add_mapped_ortholog_column() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.add_mapped_ortholog_column"]], "build_gene_mapping() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.build_gene_mapping"]], "build_ortho_mapping() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.build_ortho_mapping"]], "convert_ids() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.convert_ids"]], "convert_to_human() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.convert_to_human"]], "convert_to_zebrafish() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.convert_to_zebrafish"]], "danrerlib.mapping": [[7, "module-danrerlib.mapping"]], "get_ortho_ids() (in module danrerlib.mapping)": [[7, "danrerlib.mapping.get_ortho_ids"]], "database_dir (in module danrerlib.settings)": [[8, "danrerlib.settings.DATABASE_DIR"]], "ens_id (in module danrerlib.settings)": [[8, "danrerlib.settings.ENS_ID"]], "file_dir (in module danrerlib.settings)": [[8, "danrerlib.settings.FILE_DIR"]], "go_data_dir (in module danrerlib.settings)": [[8, "danrerlib.settings.GO_DATA_DIR"]], "human_id (in module danrerlib.settings)": [[8, "danrerlib.settings.HUMAN_ID"]], "kegg_data_dir (in module danrerlib.settings)": [[8, "danrerlib.settings.KEGG_DATA_DIR"]], "master_mapping_file_path (in module danrerlib.settings)": [[8, "danrerlib.settings.MASTER_MAPPING_FILE_PATH"]], "master_ortho_file_path (in module danrerlib.settings)": [[8, "danrerlib.settings.MASTER_ORTHO_FILE_PATH"]], "ncbi_id (in module danrerlib.settings)": [[8, "danrerlib.settings.NCBI_ID"]], "raw_data_dir (in module danrerlib.settings)": [[8, "danrerlib.settings.RAW_DATA_DIR"]], "symbol (in module danrerlib.settings)": [[8, "danrerlib.settings.SYMBOL"]], "version_num (in module danrerlib.settings)": [[8, "danrerlib.settings.VERSION_NUM"]], "zfin_id (in module danrerlib.settings)": [[8, "danrerlib.settings.ZFIN_ID"]], "danrerlib.settings": [[8, "module-danrerlib.settings"]], "invalidorganismerror": [[9, "danrerlib.utils.InvalidOrganismError"]], "check_valid_human_gene_id_type() (in module danrerlib.utils)": [[9, "danrerlib.utils.check_valid_human_gene_id_type"]], "check_valid_organism() (in module danrerlib.utils)": [[9, "danrerlib.utils.check_valid_organism"]], "check_valid_zebrafish_gene_id_type() (in module danrerlib.utils)": [[9, "danrerlib.utils.check_valid_zebrafish_gene_id_type"]], "danrerlib.utils": [[9, "module-danrerlib.utils"]], "download_url() (in module danrerlib.utils)": [[9, "danrerlib.utils.download_url"]], "download_url_to_file() (in module danrerlib.utils)": [[9, "danrerlib.utils.download_url_to_file"]], "normalize_gene_id_type() (in module danrerlib.utils)": [[9, "danrerlib.utils.normalize_gene_id_type"]], "normalize_organism_name() (in module danrerlib.utils)": [[9, "danrerlib.utils.normalize_organism_name"]], "pretty_print_series() (in module danrerlib.utils)": [[9, "danrerlib.utils.pretty_print_series"]], "save_data() (in module danrerlib.utils)": [[9, "danrerlib.utils.save_data"]]}})